<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>2014 analysis of influenza NP with stringency parameter &mdash; phyloExpCM 0.3 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="phyloExpCM 0.3 documentation" href="index.html" />
    <link rel="up" title="Examples" href="examples.html" />
    <link rel="next" title="2014 analysis of H1 influenza HAs descended from 1918 virus" href="example_2014Analysis_Influenza_H1_HA.html" />
    <link rel="prev" title="2014 analysis of beta-lactamase phylogenetic fit" href="example_2014Analysis_lactamase.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="example_2014Analysis_Influenza_H1_HA.html" title="2014 analysis of H1 influenza HAs descended from 1918 virus"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="example_2014Analysis_lactamase.html" title="2014 analysis of beta-lactamase phylogenetic fit"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">phyloExpCM 0.3 documentation</a> &raquo;</li>
          <li><a href="examples.html" accesskey="U">Examples</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="analysis-of-influenza-np-with-stringency-parameter">
<h1><a class="toc-backref" href="#id1">2014 analysis of influenza NP with stringency parameter</a><a class="headerlink" href="#analysis-of-influenza-np-with-stringency-parameter" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#analysis-of-influenza-np-with-stringency-parameter" id="id1">2014 analysis of influenza NP with stringency parameter</a><ul>
<li><a class="reference internal" href="#overview" id="id2">Overview</a></li>
<li><a class="reference internal" href="#software-used" id="id3">Software used</a></li>
<li><a class="reference internal" href="#input-files" id="id4">Input files</a></li>
<li><a class="reference internal" href="#running-the-analysis" id="id5">Running the analysis</a></li>
<li><a class="reference internal" href="#steps-in-the-analysis" id="id6">Steps in the analysis</a><ul>
<li><a class="reference internal" href="#creation-of-sequence-alignments" id="id7">Creation of sequence alignments</a></li>
<li><a class="reference internal" href="#check-for-outlier-sequences" id="id8">Check for outlier sequences</a></li>
<li><a class="reference internal" href="#construction-of-phylogenetic-tree" id="id9">Construction of phylogenetic tree</a></li>
<li><a class="reference internal" href="#optimizing-the-tree-with-various-substitution-models" id="id10">Optimizing the tree with various substitution models</a></li>
<li><a class="reference internal" href="#comparison-of-per-site-likelihoods" id="id11">Comparison of per-site likelihoods</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id2">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>This is a description of an analysis of the effectiveness of experimentally determined substitution models for describing the evolution of influenza nucleoprotein (NP).</p>
<p>This analysis closely parallels the one that has been previously described
in the paper <a class="reference external" href="http://mbe.oxfordjournals.org/content/31/8/1956">An experimentally determined evolutionary model dramatically improves phylogenetic fit</a> and is available at <a class="reference external" href="http://jbloom.github.io/phyloExpCM/example_2013Analysis_Influenza_NP_Human_1918_Descended.html">http://jbloom.github.io/phyloExpCM/example_2013Analysis_Influenza_NP_Human_1918_Descended.html</a></p>
<p>It differs from that analysis primarily in its use of a stringency parameter <span class="math">\(\beta\)</span> for the amino-acid preferences, and in its site-specific analysis. The use of this stringency parameter and the results from the analysis are described in <a class="reference external" href="http://dx.doi.org/10.1093/molbev/msu220">An experimentally informed evolutionary model improves phylogenetic fit to divergent lactamase homologs</a>.</p>
<p>This analysis was performed by <a class="reference external" href="http://research.fhcrc.org/bloom/en.html">Jesse Bloom</a>.</p>
<p>The overall approach is to use the <a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> package to analyze a NP phylogeny using an experimentally derived substitution model from a <a class="reference external" href="http://jbloom.github.io/mapmuts/example_2013Analysis_Influenza_NP_Aichi68.html">mapmuts analysis of the Aichi68 influenza NP from 2013</a>, and to compare this model to various standard codon substitution models.</p>
<p>Much of the analysis itself is performed using <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a>, which is run by <a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> using the scripts detailed in the <a class="reference external" href="http://jbloom.github.io/phyloExpCM">phyloExpCM documentation</a>.</p>
<p>The phylogenetic tree that is analyzed consists of human influenza lineages with their NPs directly descended from the 1918 virus, specifically:</p>
<blockquote>
<div><ul class="simple">
<li>H1N1 from 1918 to 1957, and again from 1977 to 2008 (seasonal H1N1).</li>
<li>H2N2 from 1957 to 1968.</li>
<li>H3N2 from 1968 to 2013.</li>
</ul>
</div></blockquote>
<p>The code, data, and results from the analysis are in the <tt class="docutils literal"><span class="pre">./examples/2014Analysis_Influenza_NP_Human_1918_Descended_withbeta/</span></tt> subdirectory in the main <a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> package directory. This subdirectory contains all of the necessary input files (see <a class="reference internal" href="#input-files">Input files</a>) for running the analysis. It also contains a master script (<tt class="docutils literal"><span class="pre">run_analysis.py</span></tt>) that can be used to regenerate the analysis with the command:</p>
<div class="highlight-python"><div class="highlight"><pre>python run_analysis.py
</pre></div>
</div>
<p>provided that you have installed the software described in <a class="reference internal" href="#software-used">Software used</a>.</p>
</div>
<div class="section" id="software-used">
<h2><a class="toc-backref" href="#id3">Software used</a><a class="headerlink" href="#software-used" title="Permalink to this headline">¶</a></h2>
<p>The analysis was performed using the following software on the Fred Hutchinson computer cluster using the following Linux distribution:</p>
<div class="highlight-python"><div class="highlight"><pre>Linux version 3.5.0-43-generic (buildd@roseapple) (gcc version 4.6.3 (Ubuntu/Linaro 4.6.3-1ubuntu5) ) #66~precise1-Ubuntu SMP Thu Oct 24 14:52:23 UTC 2013
</pre></div>
</div>
<p>In order to replicate the analysis, you will need to use Linux or Mac OS X, and install the software listed below.</p>
<p>The specific versions of software used are specified. Other versions will likely work as well, but this has not been confirmed.</p>
<ul>
<li><p class="first"><a class="reference external" href="http://www.python.org/">Python</a> (version 2.7.3)</p>
</li>
<li><p class="first"><a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> : specifically, <a class="reference external" href="https://github.com/jbloom/phyloExpCM/tree/v0.3">phyloExpCM version 0.3</a></p>
</li>
<li><p class="first"><a class="reference external" href="http://www.numpy.org/">numpy</a> (version 1.6.1)</p>
</li>
<li><p class="first"><a class="reference external" href="http://matplotlib.org/">matplotlib</a> (version 1.3.1).</p>
</li>
<li><p class="first"><a class="reference external" href="https://github.com/jbloom/mapmuts">mapmuts</a> : specifically, <a class="reference external" href="https://github.com/jbloom/mapmuts/tree/v1.0">mapmuts version 1.0</a></p>
</li>
<li><p class="first"><a class="reference external" href="http://emboss.sourceforge.net/download/">EMBOSS needle</a> (version 6.6.0). The executable with the name <tt class="docutils literal"><span class="pre">needle</span></tt> is expected to be in the current search path.</p>
</li>
<li><p class="first"><a class="reference external" href="https://github.com/stamatak/standard-RAxML">RAxML</a> (version 7.7.6). The executable with the name <tt class="docutils literal"><span class="pre">raxmlHPC-SSE3</span></tt> is expected to be in the current search path. This is the SSE3 version, and was built using the command:</p>
<div class="highlight-python"><div class="highlight"><pre>make -f Makefile.SSE3.gcc
</pre></div>
</div>
<p>and then copying the executable into the search path.</p>
</li>
<li><p class="first"><a class="reference external" href="http://tree.bio.ed.ac.uk/software/pathogen/">Path-O-Gen</a> (version 1.4) was used for the visual identification of outliers from the molecular clock and to root trees based on date-stamping.</p>
</li>
<li><p class="first"><a class="reference external" href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a> (version 1.4.0) was used to open tree files for annotation, visualization, and saving of images.</p>
</li>
<li><p class="first"><a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> (version 1.00 201306.18) was used to build phylogenetic trees. The executable with the name <tt class="docutils literal"><span class="pre">codonphyml</span></tt> is expected to be in the current search path. This executable was compiled with the <em>omp</em> options to make it multi-threaded.</p>
</li>
<li><p class="first"><a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> (version HYPHY 2.220131214beta(MP) for Linux on x86_64) was used to optimize the branches and compute the likelihoods under the different models. The multiprocessor (<tt class="docutils literal"><span class="pre">HYPHYMP</span></tt>) version of the executable was used. To install this, the following commands were run in the unzipped <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> source code directory after download:</p>
<div class="highlight-python"><div class="highlight"><pre>cmake -DINSTALL_PREFIX=/home/jbloom/.local/ .
make MP2
make install
</pre></div>
</div>
</li>
<li><p class="first">This analysis was run on the FHCRC scientific computing cluster using <a class="reference external" href="https://computing.llnl.gov/linux/slurm/sbatch.html">sbatch</a> (version 2.6.2), and requires <a class="reference external" href="https://computing.llnl.gov/linux/slurm/sbatch.html">sbatch</a> to be installed to run as currently implemented. If you don&#8217;t want to use <a class="reference external" href="https://computing.llnl.gov/linux/slurm/sbatch.html">sbatch</a>, all you need to do is modify the main <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> script to not submit any jobs via <a class="reference external" href="https://computing.llnl.gov/linux/slurm/sbatch.html">sbatch</a>. The use of <a class="reference external" href="https://computing.llnl.gov/linux/slurm/sbatch.html">sbatch</a> in the current version simply allows more processors to be used simultaneously and so speeds up the analysis.</p>
</li>
<li><p class="first"><a class="reference external" href="http://www.imagemagick.org/script/convert.php">ImageMagick convert</a> 6.6.9-7 2012-08-17 was used to convert PDF files to JPGs.</p>
</li>
</ul>
</div>
<div class="section" id="input-files">
<h2><a class="toc-backref" href="#id4">Input files</a><a class="headerlink" href="#input-files" title="Permalink to this headline">¶</a></h2>
<p>Here are the input files that are necessary to repeat the analysis. These are the names assigned with the <tt class="docutils literal"><span class="pre">./examples/2013Analysis_Influenza_NP_Human_1918_Descended/</span></tt> subdirectory of the main <a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> package.</p>
<ul>
<li><p class="first"><tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> is a <a class="reference external" href="http://www.python.org/">Python</a> script that will perform the analysis assuming that you have all of the appropriate software (see <a class="reference internal" href="#software-used">Software used</a>) and input files. To run this script, use:</p>
<div class="highlight-python"><div class="highlight"><pre>python run_analysis.py
</pre></div>
</div>
<p>at the command line.</p>
</li>
<li><p class="first"><em>NPseqs.fasta</em> is the set of all unique full-length influenza A coding DNA sequences as downloaded from the <a class="reference external" href="http://www.ncbi.nlm.nih.gov/genomes/FLU/FLU.html">Influenza Virus Resource</a> on June-25-2013.</p>
</li>
<li><p class="first"><em>Aichi68-NP.fasta</em> is the coding DNA sequence for A/Aichi/2/1968 (H3N2) NP as taken from reverse-genetics plasmid pHWAichi68-NP.</p>
</li>
<li><p class="first"><em>JVI_82_8947_Anomalies.txt</em> is a list of the strain names for the sequences identified as anomalous (either frozen in time or recombinant) in Appendices 1 and 2 of <cite>Krasnitz et al, 2008</cite>.</p>
</li>
<li><p class="first"><em>JDB_Anomalies.txt</em> is a list of strain names that appear to be anomalous based on their strong violation of the molecular clock. These sequences are probably mis-annotated in the database, come from non-standard lineages (for examples a swine H3N2 isolated from humans), are reassortants (example would be an H1N1 NP that reassorted into an H3N2), or are extensively lab adapted and so deviate from the molecular clock due to large numbers ofadaptation mutations.</p>
</li>
<li><p class="first"><em>mutspectrum.txt</em> is a file that contains the experimentally measured nucleotide mutation rates. The measurement of these mutation rates is described in <a class="reference external" href="http://mbe.oxfordjournals.org/content/31/8/1956">An experimentally determined evolutionary model dramatically improves phylogenetic fit</a>.</p>
</li>
<li><p class="first">File giving the inferred equilibrium amino-acid preferences at each site in NP. This file is from the <a class="reference external" href="http://jbloom.github.io/mapmuts/example_2013Analysis_Influenza_NP_Aichi68.html">mapmuts analysis of the Aichi68 influenza NP from 2013</a>. The file is <tt class="docutils literal"><span class="pre">p1_equilibriumpreferences.txt</span></tt>. It is the combined experimental data from the two passage 1 replicates described in <a class="reference external" href="http://mbe.oxfordjournals.org/content/31/8/1956">An experimentally determined evolutionary model dramatically improves phylogenetic fit</a>.</p>
</li>
<li><p class="first"><tt class="docutils literal"><span class="pre">./DSSP_analysis/</span></tt> is a subdirectory that contains the results of analyzing the solvent accessibilities and secondary structures of NP from the crystal structure. Briefly, the NP crystal structure in <a class="reference external" href="http://www.rcsb.org/pdb/explore.do?structureId=2iqh">PDB 2IQH</a> was downloaded, and the <tt class="docutils literal"><span class="pre">2IQH.pdb</span></tt> file was manually edited to extract only monomer C into <tt class="docutils literal"><span class="pre">2IQH_monomerC.pdb</span></tt>. This monomer was then analyzed with the <a class="reference external" href="http://www.cmbi.ru.nl/hsspsoap/">DSSP webserver</a> to create the results stored in <tt class="docutils literal"><span class="pre">2IQH_monomerC.dssp</span></tt>.</p>
</li>
</ul>
</div>
<div class="section" id="running-the-analysis">
<h2><a class="toc-backref" href="#id5">Running the analysis</a><a class="headerlink" href="#running-the-analysis" title="Permalink to this headline">¶</a></h2>
<p>The analysis in this directory can be repeated simply by running the master script <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> from the command line with:</p>
<div class="highlight-python"><div class="highlight"><pre>python run_analysis.py
</pre></div>
</div>
<p>or running it using <a class="reference external" href="https://computing.llnl.gov/linux/slurm/sbatch.html">sbatch</a> with:</p>
<div class="highlight-python"><div class="highlight"><pre>sbatch run_analysis.sbatch
</pre></div>
</div>
<p>Other than the files delineated under <a class="reference internal" href="#input-files">Input files</a>, all data files in this directory can be re-generated with this command. However, the analysis approach does involve a few manual steps with <a class="reference external" href="http://tree.bio.ed.ac.uk/software/pathogen/">Path-O-Gen</a> and <a class="reference external" href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a> to create nicely formatted and annotated tree images from the raw tree files produced by the script. These manual steps are indicated below.</p>
<p>There is a variable called <em>use_existing_output</em> that is hardcoded into the beginning of this script. This variable is currently set to <em>True</em>, which means that if output already exists it is <strong>not</strong> overwritten and regenerated. If you want to regenerate the output (such as after changing input files or altering analysis parameters), then either manually delete the old output or change <em>use_existing_output</em> to <em>False</em> to force the script to overwrite old output. This script does <strong>not</strong> automatically test for dependencies among files, so if you change a file but keep <em>use_existing_output</em> to <em>True</em>, then you need to manually remove any old files with dependencies.</p>
<p>Note that <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> uses multiple CPUs to speed up some of the analyses. If you don&#8217;t have as many CPUs on your system, you might want to check that you are not overloading the system &#8211; you can avoid this by using fewer CPUs, which will require you to manually modify the <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> script.</p>
<p>Note that <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> uses <a class="reference external" href="https://computing.llnl.gov/linux/slurm/sbatch.html">sbatch</a> by default to run some of its subsidiary programs &#8211; if you don&#8217;t want to do this, set the <em>use_sbatch</em> option to <em>False</em> within the program.</p>
</div>
<div class="section" id="steps-in-the-analysis">
<h2><a class="toc-backref" href="#id6">Steps in the analysis</a><a class="headerlink" href="#steps-in-the-analysis" title="Permalink to this headline">¶</a></h2>
<p>Here are the steps in the analysis. The command for each step is shown, but these commands (with the exception of some manual formatting of tree images with <a class="reference external" href="http://tree.bio.ed.ac.uk/software/pathogen/">Path-O-Gen</a> and <a class="reference external" href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a>) are run automatically by the main <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> script if you are using that.</p>
<div class="section" id="creation-of-sequence-alignments">
<h3><a class="toc-backref" href="#id7">Creation of sequence alignments</a><a class="headerlink" href="#creation-of-sequence-alignments" title="Permalink to this headline">¶</a></h3>
<p>Extract and align the human NP sequences descended from the 1918 virus into <em>Aligned_NPs.fasta</em>. This is done by the script <tt class="docutils literal"><span class="pre">get_treeseqs.py</span></tt>, by running:</p>
<div class="highlight-python"><div class="highlight"><pre>python get_treeseqs.py
</pre></div>
</div>
<p>The created <em>Aligned_NPs.fasta</em> includes the sequences with headers that give their isolation date (except for H1N1 from after 1977, where 24 years are subtracted from the isolation date for the reasons described below).</p>
<p>The specific steps are as follows:</p>
<ol class="arabic">
<li><p class="first">Read all sequences from <em>NPseqs.fasta</em>.</p>
</li>
<li><p class="first">Purge any sequences specified as anomalous in <em>JVI_82_8947_Anomalies.txt</em> and <em>JDB_Anomalies.txt</em>.</p>
</li>
<li><p class="first">Purge any sequences with ambiguous nucleotide identities.</p>
</li>
<li><p class="first">If there are redundant identical sequences, keep only one sequence from each set of identical sequences.</p>
</li>
<li><p class="first">Purge any sequences for which the host, year of isolation, or subtype is unavailable.</p>
</li>
<li><p class="first">Purge any sequences for which the host is something other than human.</p>
</li>
<li><p class="first">Keep at most 3 sequences per year from each of the following classifications. Sequences from other classifications are all removed. The classifications for which 3 per year are retained are:</p>
<blockquote>
<div><ul class="simple">
<li>H1N1 from 1918 to 1957: this is H1N1 directly descended from the 1918 virus.</li>
<li>H2N2 from 1957 to 1968: this H2N2 derived from reassortment of an avian virus with the existing H1N1 in the 1957 &#8220;Asian flu&#8221; pandemic. The NP in the new H2N2 was from the existing H1N1, which then went extinct.</li>
<li>H3N2 from 1968 to 2013: this is H3N2 derived from reassortment of an avian virus with thte existing H2N2 in the 1968 &#8220;Hong Kong flu&#8221; pandemic. The NP in the new H3N2 was from the existing H2N2, when then went extinct.</li>
<li>H1N1 from 1977 to 2008: this is &#8220;seasonal H1N1&#8221; that arose in 1977 in the &#8220;Russian flu&#8221; pandemic. This is a virus from the earlier human H1N1 that was frozen and re-released. The virus was frozen for about 24 years before this pandemic (estimate from <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/19787384">dos Reis et al, 2009</a>), and therefore <strong>24 years are subtracted from the isolation date of viruses in this lineage</strong>. This date subtraction is not important for the construction of the phylogenetic tree, but it is important for the <a class="reference external" href="http://tree.bio.ed.ac.uk/software/pathogen/">Path-O-Gen</a> based analysis that assumes a molecular clock (see <a class="reference internal" href="#check-for-outlier-sequences">Check for outlier sequences</a>).</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Translate and align all of the protein sequences, then also construct nucleotide alignments from the protein alignments. The alignments are built with <a class="reference external" href="http://emboss.sourceforge.net/download/">EMBOSS needle</a>. The alignments are written to <em>Aligned_NPs.fasta</em> and <em>Aligned_NP_proteins.fasta</em>.</p>
</li>
</ol>
</div>
<div class="section" id="check-for-outlier-sequences">
<h3><a class="toc-backref" href="#id8">Check for outlier sequences</a><a class="headerlink" href="#check-for-outlier-sequences" title="Permalink to this headline">¶</a></h3>
<p>There are almost always clear &#8220;outlier&#8221; sequences in influenza phylogenies. These outliers can arise from several sources:</p>
<blockquote>
<div><ul class="simple">
<li>Mis-annotation of the date or subtype in the database.</li>
<li>Amplification of a lab contaminant during sequencing.</li>
<li>Sequencing of a zoonotic rather than epidemic strain. For example, there are currently low rates of swine-origin vH3N2 in humans. These strains are also sequenced and labeld as human H3N2 (since they are H3N2 from humans) &#8211; but they are not from the same evolutionary lineage, as they have been evolving in swine.</li>
</ul>
</div></blockquote>
<p>To eliminate such outliers, several steps were taken. First, sequences that have already been identified as anomalous <a class="reference external" href="http://jvi.asm.org/content/82/17/8947.abstract">Krasnitz et al, 2008</a> are eliminated in <a class="reference internal" href="#creation-of-sequence-alignments">Creation of sequence alignments</a> described above.</p>
<p>However, to identify further outliers, the following steps were taken. Essentially, these steps utilize <a class="reference external" href="https://github.com/stamatak/standard-RAxML">RAxML</a> and <a class="reference external" href="http://tree.bio.ed.ac.uk/software/pathogen/">Path-O-Gen</a> to manually identify sequences that are clear outliers from the molecular clock. These sequences are then added to the file <em>JDB_Anomalies</em>, and the steps in <a class="reference internal" href="#creation-of-sequence-alignments">Creation of sequence alignments</a> and then those here are repeated. This iterative process has been completed for the sequences here until no more outliers were identified.</p>
<p>Specifically, the steps were:</p>
<ol class="arabic">
<li><p class="first">The <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt> script builds a <a class="reference external" href="https://github.com/stamatak/standard-RAxML">RAxML</a> tree for quick visual inspection. <a class="reference external" href="https://github.com/stamatak/standard-RAxML">RAxML</a> is used because it is a rapid method to build a reasonable maximum-likelihood tree. The steps:</p>
<blockquote>
<div><ol class="loweralpha">
<li><p class="first">The subdirectory <tt class="docutils literal"><span class="pre">./RAxML_output/</span></tt> was created</p>
</li>
<li><p class="first"><a class="reference external" href="https://github.com/stamatak/standard-RAxML">RAxML</a> was run using the following command:</p>
<div class="highlight-python"><div class="highlight"><pre>raxmlHPC-SSE3 -w ./RAxML_output -n Aligned_NPs -p 1 -m GTRCAT -s Aligned_NPs.fasta
</pre></div>
</div>
<p>Note that for the actual command, the full path needs to be specified for the <tt class="docutils literal"><span class="pre">RAxML_output</span></tt> subdirectory as <a class="reference external" href="https://github.com/stamatak/standard-RAxML">RAxML</a> does not handle relative paths &#8211; this is handled by the master script <tt class="docutils literal"><span class="pre">run_analysis.py</span></tt>.</p>
</li>
<li><p class="first">The following command will have created the best <a class="reference external" href="https://github.com/stamatak/standard-RAxML">RAxML</a> tree as <tt class="docutils literal"><span class="pre">./RAxML_output/RAxML_bestTree.Aligned_NPs</span></tt>.</p>
</li>
</ol>
</div></blockquote>
</li>
<li><p class="first">The next step is manual. Open this tree using <a class="reference external" href="http://tree.bio.ed.ac.uk/software/pathogen/">Path-O-Gen</a>, and look manually to see if there are outliers from the molecular clock in the <tt class="docutils literal"><span class="pre">./RAxML_output/RAxML_bestTree.Aligned_NPs</span></tt> file. If so, add these outliers to <em>JDB_Anomalies.txt</em> and repeat the <a class="reference internal" href="#creation-of-sequence-alignments">Creation of sequence alignments</a> and this <a class="reference external" href="https://github.com/stamatak/standard-RAxML">RAxML</a> analysis. Note that this has already been done, so if you are just re-running the analysis then you don&#8217;t need to do this unless you have changed the sequence set.</p>
</li>
</ol>
</div>
<div class="section" id="construction-of-phylogenetic-tree">
<h3><a class="toc-backref" href="#id9">Construction of phylogenetic tree</a><a class="headerlink" href="#construction-of-phylogenetic-tree" title="Permalink to this headline">¶</a></h3>
<p>A high-quality phylogenetic trees were then constructed using <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a>, which is able to build maximum-likelihood trees with codon substitution models.</p>
<p>The <a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> script <tt class="docutils literal"><span class="pre">phyloExpCM_runcodonPhyML.py</span></tt> (see <a class="reference external" href="http://jbloom.github.io/phyloExpCM">phyloExpCM documentation</a> for details) was used to run <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a>. Trees were built using
the <em>GY94</em> codon model (<a class="reference external" href="http://mbe.oxfordjournals.org/content/11/5/725.full.pdf">Goldman and Yang 1994</a>) with:</p>
<blockquote>
<div><ul class="simple">
<li>The equilibrium codon frequencies estimated emprically using the <a class="reference external" href="http://www.plosone.org/article/info%3Adoi/10.1371/journal.pone.0011230">CF3x4</a> method.</li>
<li>A single transition / transversion ratio (<em>kappa</em>) estimated by maximum likelihood.</li>
<li>The dN/dS ratio (<em>omega</em>) drawn from four discrete gamma-distributed categories (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/7932792">Yang 1994</a>) with the shape parameter and mean estimated by maximum likelihood.</li>
</ul>
</div></blockquote>
<p>The tree construction was performed in the <tt class="docutils literal"><span class="pre">./GY94_codonPhyML_tree/</span></tt> subdirectory. The tree itself is in the file <tt class="docutils literal"><span class="pre">./GY94_codonPhyML_tree/codonphyml_tree.newick</span></tt>.</p>
</div>
<div class="section" id="optimizing-the-tree-with-various-substitution-models">
<h3><a class="toc-backref" href="#id10">Optimizing the tree with various substitution models</a><a class="headerlink" href="#optimizing-the-tree-with-various-substitution-models" title="Permalink to this headline">¶</a></h3>
<p>This is the heart of the analysis: to use various codon substitution models to optimize the branch lengths and likelihood for the tree topology estimated with <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> and the <em>GY94</em> substitution model as described in <a class="reference internal" href="#construction-of-phylogenetic-tree">Construction of phylogenetic tree</a>. This analysis compares the experimentally determined codon substitution models with various other substitution models.</p>
<p>This analysis is done with the <a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> scripts <tt class="docutils literal"><span class="pre">phyloExpCM_ExpModelOptimizeHyphyTree.py</span></tt> and <tt class="docutils literal"><span class="pre">phyloExpCM_optimizeHyphyTree.py</span></tt>, which are described in the <a class="reference external" href="http://jbloom.github.io/phyloExpCM">phyloExpCM documentation</a>. These scripts uses <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> to optimize the branch lengths and any substitution model free parameters on the fixed tree topology estimated with <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a>.</p>
<p>The results of the optimizations for the various substitution models are in the subdirectory <tt class="docutils literal"><span class="pre">./codonmodel_optimized_trees/</span></tt>. Within this subdirectory are further subdirectories containing the results of optimizing the branch lengths with <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> using the indicated codon substitution models. For example, <tt class="docutils literal"><span class="pre">./codonmodel_optimized_trees/Tree-GY94_Model-KOSI07_F_omega-global-gamma4_rates-one/</span></tt> contains one such model. Within this subdirectory, the files of interest are:</p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">hyphy_tree.newick</span></tt> or <tt class="docutils literal"><span class="pre">optimizedtree.newick</span></tt> (file name differs for different analyses) is the tree with branch lengths optimized by <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> using the substitution model in question.</li>
<li><tt class="docutils literal"><span class="pre">hyphy_output.txt</span></tt> or <tt class="docutils literal"><span class="pre">optimizedtree_results.txt</span></tt> (file name differs for different analyses) is a text file summarizing the <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> output from the optimization.</li>
<li><tt class="docutils literal"><span class="pre">sitelikelihoods.txt</span></tt> gives the per-site likelihoods.</li>
</ul>
</div></blockquote>
<p>A more helpful way to view the results is to look at the main summary files created in the examples home directory. These files are <tt class="docutils literal"><span class="pre">GY94_summary.csv</span></tt> (for optimization of the tree topology estimated by <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> with the <em>GY94</em> model) and a <tt class="docutils literal"><span class="pre">LaTex</span></tt> version of this file called <tt class="docutils literal"><span class="pre">GY94_summary.tex</span></tt>.</p>
<p>Here are the results for optimization of the tree topology inferred with the <em>GY94</em> model in <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> (this is the file <tt class="docutils literal"><span class="pre">GY94_summary.csv</span></tt>):</p>
<blockquote>
<div><div class="highlight-python"><div class="highlight"><pre>#Summary for tree GY94.
#
#SUBSTITUTION_MODEL, dAIC, LOG_LIKELIHOOD, FREE_PARAMETERS, MAXIMUM_LIKELIHOOD_PARAMETERS, EMPIRICAL_PARAMETERS
fitbetaFracTolerated, 0, -12144.2, 1, 1, 0
FracTolerated, 391.8, -12341.1, 0, 0, 0
GY94_CF3x4_omega-global-gamma4_rates-gamma4, 1453.03, -12858.7, 13, 4, 9
GY94_CF3x4_omega-global-gamma4_rates-one, 1616, -12941.2, 12, 3, 9
KOSI07_F_omega-global-gamma4_rates-gamma4, 1845.6, -13004, 64, 4, 60
GY94_CF3x4_omega-global-one_rates-gamma4, 1884.13, -13075.3, 12, 3, 9
GY94_CF3x4_omega-global-one_rates-one, 2153.21, -13210.8, 11, 2, 9
KOSI07_F_omega-global-gamma4_rates-one, 2153.56, -13159, 63, 3, 60
KOSI07_F_omega-global-one_rates-gamma4, 2227.3, -13195.9, 63, 3, 60
KOSI07_F_omega-global-one_rates-one, 2650.12, -13408.3, 62, 2, 60
avgaafreqsfitbeta_FracTolerated, 3736.2, -13952.3, 61, 1, 60
avgaafreqs_FracTolerated, 3742.2, -13956.3, 60, 0, 60
</pre></div>
</div>
</div></blockquote>
<p>These results show that inclusion of the stringency parameter <span class="math">\(\beta\)</span> improves the phylogenetic fit.</p>
</div>
<div class="section" id="comparison-of-per-site-likelihoods">
<h3><a class="toc-backref" href="#id11">Comparison of per-site likelihoods</a><a class="headerlink" href="#comparison-of-per-site-likelihoods" title="Permalink to this headline">¶</a></h3>
<p>The <cite>phyloExpCM</cite> script <tt class="docutils literal"><span class="pre">phyloExpCM_SiteLikelihoodComparison.py</span></tt> was used to compare the per-site likelihoods between the best experimentally informed model (the <em>HalpernBruno</em> model with a fitted <span class="math">\(\beta\)</span>) and the best traditional model (the <em>GY94</em>
model with gamma-distributed rates and <span class="math">\(\omega\)</span>).</p>
<p>This was done by creating the input file <tt class="docutils literal"><span class="pre">phyloExpCM_SitelikelihoodComparison_infile.txt</span></tt>
and then running:</p>
<div class="highlight-python"><div class="highlight"><pre>phyloExpCM_SiteLikelihoodComparison.py phyloExpCM_SiteLikelihoodComparison_infile.txt
</pre></div>
</div>
<p>This created three output files.</p>
<p>The first, <tt class="docutils literal"><span class="pre">sitelikelihoods.txt</span></tt> lists the differences in per-site likelihoods by site.</p>
<p>The next two are the following plots, which use Tukey box plots to show the differences in site likelihoods, categorizing sites by the <a class="reference external" href="http://www.cmbi.ru.nl/hsspsoap/">DSSP webserver</a> computed secondary
structures (<em>SS</em>) and relative solvent accessibilities (<em>RSA</em>) in <tt class="docutils literal"><span class="pre">DSSP_analysis/2IQH_monomerC.dssp</span></tt>
Here are these plots:</p>
<div class="figure align-center">
<a class="reference internal image-reference" href="_images/example_2014Analysis_Influenza_NP_Human_1918_Descended_withbeta_sitelikelihoodcomparison_bySS.jpg"><img alt="sitelikelihoodcomparison_bySS.jpg" src="_images/example_2014Analysis_Influenza_NP_Human_1918_Descended_withbeta_sitelikelihoodcomparison_bySS.jpg" style="width: 45%;" /></a>
<p class="caption">The comparison of site likelihoods with sites categorized by secondary structure (<tt class="docutils literal"><span class="pre">sitelikelihoodcomparison_bySS.pdf</span></tt>).</p>
</div>
<div class="figure align-center">
<a class="reference internal image-reference" href="_images/example_2014Analysis_Influenza_NP_Human_1918_Descended_withbeta_sitelikelihoodcomparison_byRSA.jpg"><img alt="sitelikelihoodcomparison_byRSA.jpg" src="_images/example_2014Analysis_Influenza_NP_Human_1918_Descended_withbeta_sitelikelihoodcomparison_byRSA.jpg" style="width: 45%;" /></a>
<p class="caption">The comparison of site likelihoods with sites categorized by relative solvent accessibility (<tt class="docutils literal"><span class="pre">sitelikelihoodcomparison_byRSA.pdf</span></tt>).</p>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">2014 analysis of influenza NP with stringency parameter</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#software-used">Software used</a></li>
<li><a class="reference internal" href="#input-files">Input files</a></li>
<li><a class="reference internal" href="#running-the-analysis">Running the analysis</a></li>
<li><a class="reference internal" href="#steps-in-the-analysis">Steps in the analysis</a><ul>
<li><a class="reference internal" href="#creation-of-sequence-alignments">Creation of sequence alignments</a></li>
<li><a class="reference internal" href="#check-for-outlier-sequences">Check for outlier sequences</a></li>
<li><a class="reference internal" href="#construction-of-phylogenetic-tree">Construction of phylogenetic tree</a></li>
<li><a class="reference internal" href="#optimizing-the-tree-with-various-substitution-models">Optimizing the tree with various substitution models</a></li>
<li><a class="reference internal" href="#comparison-of-per-site-likelihoods">Comparison of per-site likelihoods</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="example_2014Analysis_lactamase.html"
                        title="previous chapter">2014 analysis of beta-lactamase phylogenetic fit</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="example_2014Analysis_Influenza_H1_HA.html"
                        title="next chapter">2014 analysis of H1 influenza HAs descended from 1918 virus</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/example_2014Analysis_Influenza_NP_Human_1918_Descended_withbeta.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="example_2014Analysis_Influenza_H1_HA.html" title="2014 analysis of H1 influenza HAs descended from 1918 virus"
             >next</a> |</li>
        <li class="right" >
          <a href="example_2014Analysis_lactamase.html" title="2014 analysis of beta-lactamase phylogenetic fit"
             >previous</a> |</li>
        <li><a href="index.html">phyloExpCM 0.3 documentation</a> &raquo;</li>
          <li><a href="examples.html" >Examples</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2013, Jesse D. Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>