<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>phyloExpCM_runcodonPhyML.py &mdash; phyloExpCM 0.3 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="phyloExpCM 0.3 documentation" href="index.html" />
    <link rel="up" title="Scripts" href="scripts.html" />
    <link rel="next" title="phyloExpCM_buildHyphyExpCM.py" href="phyloExpCM_buildHyphyExpCM.html" />
    <link rel="prev" title="Scripts" href="scripts.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="phyloExpCM_buildHyphyExpCM.html" title="phyloExpCM_buildHyphyExpCM.py"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="scripts.html" title="Scripts"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">phyloExpCM 0.3 documentation</a> &raquo;</li>
          <li><a href="scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="phyloexpcm-runcodonphyml-py">
<span id="id1"></span><h1>phyloExpCM_runcodonPhyML.py<a class="headerlink" href="#phyloexpcm-runcodonphyml-py" title="Permalink to this headline">¶</a></h1>
<p>This script allows you to estimate a tree topology using <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a>. It has been tested with <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> version 1.00 201306.18.</p>
<p>To run the script, create an input file of the format described below. Then run the script followed by the input file, as in:</p>
<div class="highlight-python"><div class="highlight"><pre>phyloExpCM_runcodonPhyML.py infile.txt
</pre></div>
</div>
<div class="section" id="format-of-the-input-file">
<h2>Format of the input file<a class="headerlink" href="#format-of-the-input-file" title="Permalink to this headline">¶</a></h2>
<p>The input file specifies <cite>key</cite> / <cite>value</cite> pairs on individual lines in the format:</p>
<div class="highlight-python"><div class="highlight"><pre>key1 value1
key2 value2
</pre></div>
</div>
<p>Blank lines or lines
that begin with # are ignored (i.e. as comment lines). The input file should
contain the following keys:</p>
<ul>
<li><p class="first"><em>seqfile</em> should specify the name of a FASTA file with the sequences that we want to analyze. Each must have a unique header of less than 100 characters. These characters should not include any spaces or any of the following: <tt class="docutils literal"><span class="pre">:(),&quot;</span></tt> &#8211; an exception will be raised if these conditions are not met. In addition, the character <tt class="docutils literal"><span class="pre">*</span></tt> has special meaning if it is found at the end of a header (it specifies the sequence is an outgroup), so only include this if you know what you are doing. The sequences are assumed to be aligned coding sequences &#8211; an exception will be raised if they are not all of the same length.</p>
</li>
<li><p class="first"><em>codonPhyML_path</em> should specify the path to the <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> program. For example, this might just be <tt class="docutils literal"><span class="pre">codonphyml</span></tt> if the program is installed in the search path. Or it could be an absolute path name if the executable is not installed in the search path.</p>
</li>
<li><p class="first"><em>seed</em> specifies the integer random number seed passed to <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a>. Setting the same seed for identical runs should guarantee identical output.</p>
</li>
<li><p class="first"><em>model</em> specifies the codon substitution model to use with <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a>. Valid values are:</p>
<blockquote>
<div><ul class="simple">
<li><em>GY94_CF3x4</em> : The <a class="reference external" href="http://mbe.oxfordjournals.org/content/11/5/725.full.pdf">Goldman and Yang 1994</a> codon model with the equilibrium codon frequencies estimated empirically using the <a class="reference external" href="http://www.plosone.org/article/info%3Adoi/10.1371/journal.pone.0011230">CF3x4</a> method, with kappa (the transition / transversion ratio) estimated by maximum likelihood, and with omega (the dN/dS ratio) estimated as a single parameter by maximum likelihood.</li>
<li><em>GY94_CF3x4_omega-gamma4</em> : The <a class="reference external" href="http://mbe.oxfordjournals.org/content/11/5/725.full.pdf">Goldman and Yang 1994</a> codon model with the equilibrium codon frequencies estimated empirically using the <a class="reference external" href="http://www.plosone.org/article/info%3Adoi/10.1371/journal.pone.0011230">CF3x4</a> method, with kappa (the transition / transversion ratio) estimated by maximum likelihood, and with omega (the dN/dS ratio) drawn from four gamma-distributed categories with the distribution shape parameter and mean estimated by maximum likelihood.</li>
<li><em>KOSI07_F_omega-gamma4</em> : The <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/17400572">Kosiol et al, 2007</a> empirical codon model with the equilibrium codon frequencies estimated empirically using the <em>F</em> method, with <em>kappa(tv)</em> (the relative decrease in transversions versus transitions) estimated by maximum likelihood, and with <em>omega</em> (the elevation in nonsynonymous over synonymous) drawn from four gamma-distributed categories with the distribution shape parameter and mean estimated by maximum likelihood. This is based on the <em>ECM+F+omega+1kappa(tv)</em> model described by <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/17400572">Kosiol et al, 2007</a>.</li>
</ul>
</div></blockquote>
</li>
<li><p class="first"><em>outprefix</em> specifies the prefix for the names of files created by this script. Any existing files with these names are overwritten. The names and contents of theese files are specified in the output files section below.</p>
</li>
</ul>
</div>
<div class="section" id="example-input-file">
<h2>Example input file<a class="headerlink" href="#example-input-file" title="Permalink to this headline">¶</a></h2>
<p>Here is an example input file:</p>
<div class="highlight-python"><div class="highlight"><pre># Example input file for phyloExpCM_runcodonPhyML.py
seqfile Human_NPs.fasta
codonPhyML_path codonphyml
seed 1
outprefix Human_NPs_codonphyml
model GY94_CF3x4_omega-gamma4
</pre></div>
</div>
</div>
<div class="section" id="output-files">
<h2>Output files<a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>The output is as follows:</p>
<ul>
<li><p class="first">Basic progress of the script is written to standard output.</p>
</li>
<li><p class="first">Files are created with the prefix <em>outprefix</em> and the following suffixes:</p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">_config.drw</span></tt> is the darwin format input file used to run <tt class="docutils literal"><span class="pre">codonphyml</span></tt>.</li>
<li><tt class="docutils literal"><span class="pre">_output.txt</span></tt> is the output of <tt class="docutils literal"><span class="pre">codonphyml</span></tt>.</li>
<li><tt class="docutils literal"><span class="pre">_tree.newick</span></tt> is the inferred tree in Newick format. This tree contains branch supports calculated using the <a class="reference external" href="http://sysbio.oxfordjournals.org/content/60/5/685.long">SH-aLRT</a> method.</li>
<li><tt class="docutils literal"><span class="pre">_stats.txt</span></tt> contains the statistics from the inference, including all parameter values.</li>
<li><tt class="docutils literal"><span class="pre">_loglikelihood.txt</span></tt> contains a single number representing log likelihood.</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">phyloExpCM_runcodonPhyML.py</a><ul>
<li><a class="reference internal" href="#format-of-the-input-file">Format of the input file</a></li>
<li><a class="reference internal" href="#example-input-file">Example input file</a></li>
<li><a class="reference internal" href="#output-files">Output files</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="scripts.html"
                        title="previous chapter">Scripts</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="phyloExpCM_buildHyphyExpCM.html"
                        title="next chapter">phyloExpCM_buildHyphyExpCM.py</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/phyloExpCM_runcodonPhyML.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="phyloExpCM_buildHyphyExpCM.html" title="phyloExpCM_buildHyphyExpCM.py"
             >next</a> |</li>
        <li class="right" >
          <a href="scripts.html" title="Scripts"
             >previous</a> |</li>
        <li><a href="index.html">phyloExpCM 0.3 documentation</a> &raquo;</li>
          <li><a href="scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2013, Jesse D. Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>