<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>phyloExpCM_ExpModelOptimizeHyphyTree.py &mdash; phyloExpCM 0.3 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="phyloExpCM 0.3 documentation" href="index.html" />
    <link rel="up" title="Scripts" href="scripts.html" />
    <link rel="next" title="phyloExpCM_FreqsFromAlignment.py" href="phyloExpCM_FreqsFromAlignment.html" />
    <link rel="prev" title="phyloExpCM_optimizeHyphyTree.py" href="phyloExpCM_optimizeHyphyTree.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="phyloExpCM_FreqsFromAlignment.html" title="phyloExpCM_FreqsFromAlignment.py"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="phyloExpCM_optimizeHyphyTree.html" title="phyloExpCM_optimizeHyphyTree.py"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">phyloExpCM 0.3 documentation</a> &raquo;</li>
          <li><a href="scripts.html" accesskey="U">Scripts</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="phyloexpcm-expmodeloptimizehyphytree-py">
<span id="id1"></span><h1><a class="toc-backref" href="#id2">phyloExpCM_ExpModelOptimizeHyphyTree.py</a><a class="headerlink" href="#phyloexpcm-expmodeloptimizehyphytree-py" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#phyloexpcm-expmodeloptimizehyphytree-py" id="id2">phyloExpCM_ExpModelOptimizeHyphyTree.py</a><ul>
<li><a class="reference internal" href="#overview" id="id3">Overview</a></li>
<li><a class="reference internal" href="#substitution-models" id="id4">Substitution models</a><ul>
<li><a class="reference internal" href="#substitution-model" id="id5">Substitution model</a></li>
<li><a class="reference internal" href="#mutation-rates" id="id6">Mutation rates</a><ul>
<li><a class="reference internal" href="#general-model" id="id7">General model</a></li>
<li><a class="reference internal" href="#reversible-model" id="id8">Reversible model</a></li>
</ul>
</li>
<li><a class="reference internal" href="#amino-acid-preferences" id="id9">Amino-acid preferences</a></li>
<li><a class="reference internal" href="#fixation-probabilities" id="id10">Fixation probabilities</a></li>
<li><a class="reference internal" href="#equilibrium-frequencies-and-reversibility" id="id11">Equilibrium frequencies and reversibility</a></li>
</ul>
</li>
<li><a class="reference internal" href="#requirements" id="id12">Requirements</a></li>
<li><a class="reference internal" href="#running-the-script" id="id13">Running the script</a></li>
<li><a class="reference internal" href="#input-file" id="id14">Input file</a></li>
<li><a class="reference internal" href="#example-input-file" id="id15">Example input file</a></li>
<li><a class="reference internal" href="#output-files" id="id16">Output files</a><ul>
<li><a class="reference internal" href="#primary-output-files" id="id17">Primary output files</a></li>
<li><a class="reference internal" href="#temporary-output-files" id="id18">Temporary output files</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id3">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>This script uses codon substitution models derived from experimental measurements to optimize a tree of known topology. It can use either fully experimentally determined evolutionary models (known amino-acid preferences and known mutation rates), or models for which only the amino-acid preferences are known (and the mutation rates are treated as free parameters).</p>
<p>For most purposes, this script should supersede <a class="reference internal" href="phyloExpCM_optimizeHyphyTree.html#phyloexpcm-optimizehyphytree-py"><em>phyloExpCM_optimizeHyphyTree.py</em></a> for optimizing trees with experimentally determined amino-acid preferences. For experimentally determined models, it has all of the functionality of that earlier script plus more.</p>
<p>You must provide the script with a phylogenetic tree topology and a sequence alignment, as well as a set of site-specific amino-acid preferences. You can also provide a set of mutation rates, or these mutation rate parameters are optimized as free parameters (in either case, the mutation rates are shared among all sites). The amino-acid preferences and mutation rates define a codon-substitution model, which is then used to optimize the branches of the phylogenetic tree (and also the mutation rates if these are not pre-specified from experimental measurements).</p>
<p><a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> is used to perform the core analyses, and this script primarily runs <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> and manipulates the input / output.</p>
</div>
<div class="section" id="substitution-models">
<h2><a class="toc-backref" href="#id4">Substitution models</a><a class="headerlink" href="#substitution-models" title="Permalink to this headline">¶</a></h2>
<p>The model used here is the one described in <a class="reference external" href="http://dx.doi.org/10.1093/molbev/msu220">An experimentally informed evolutionary model improves phylogenetic fit to divergent lactamase homologs</a>. It may be helpful to read that reference before reading the rest of this section.</p>
<div class="section" id="substitution-model">
<h3><a class="toc-backref" href="#id5">Substitution model</a><a class="headerlink" href="#substitution-model" title="Permalink to this headline">¶</a></h3>
<p>The codon substitution models used here are defined based on experimentally measured parameters. The substitution model is determined by site-specific amino-acid preferences plus mutation rate parameters that are shared across all sites. All selection is assumed to be at the amino-acid level (selection on synonymous sites is not included). So substitution is entirely determined by knowing which mutations arise on the nucleotide level, and then how selection acts on any resulting amino-acid change.</p>
<p>In this model, the rate of substitution <span class="math">\(P_{r,xy}\)</span> from codon <em>x</em> to codon <em>y</em> is defined as</p>
<div class="math" id="equation-Prxy">
<span class="eqno">(1)</span>\[P_{r,xy} =
Q_{xy} \times F_{r,xy}\]</div>
<p>where <span class="math">\(Q_{xy}\)</span> is the rate of mutation from <em>x</em> to <em>y</em> and <span class="math">\(F_{r,xy}\)</span> is the probability that a mutation from <em>x</em> to <em>y</em> fixes once it arises. Below we describe how these mutation rates and fixation probabilities are determined. Note that this equation assumes that mutation rates are identical for all sites, but that the fixation probabilities are site specific (but site independent).</p>
</div>
<div class="section" id="mutation-rates">
<h3><a class="toc-backref" href="#id6">Mutation rates</a><a class="headerlink" href="#mutation-rates" title="Permalink to this headline">¶</a></h3>
<p>The codon mutation rates <span class="math">\(Q_{xy}\)</span> are the result of nucleotide mutations.</p>
<p>The nucleotide mutation rates can either be specified based on experimental measurements, or treated as free parameters that are optimized from the sequence data. The nucleotide mutation rates are assumed to be shared among all sites. Let <span class="math">\(R_{m \rightarrow n}\)</span> denote the rate of mutation from nucleotide <em>m</em> to <em>n</em> (more specifically, it is the probability that a nucleotide mutates from <em>m</em> to <em>n</em> in a small unit of time given that the nucleotide is already <em>m</em>), and let <span class="math">\(m_c\)</span> denote the complement of nucleotide <em>m</em> (so for example <span class="math">\(T_c = A\)</span>).</p>
<p>We assume that codon mutations happen a single nucleotide at a time, so that the <span class="math">\(Q_{xy}\)</span> in Equation <a href="#equation-Prxy">(1)</a> are given by</p>
<div class="math" id="equation-Qxy">
<span class="eqno">(2)</span>\[\begin{split}Q_{xy} =
\begin{cases}
0 &amp; \mbox{if $x$ and $y$ differ by more than on nucleotide} \\
R_{m \rightarrow n} &amp; \mbox{if $x$ differs from $y$ by a single-nucleotide change of $m$ to $n$}.
\end{cases}\end{split}\]</div>
<p>Most standard approaches to modeling nucleotide substitution (for example, the <em>GTR</em> model) assume some set of &#8220;equilibrium&#8221; nucleotide frequencies. Here we avoid taking such an approach, because these &#8220;equilibrium&#8221; frequencies lack a clear biological meaning since nucleotide frequencies are ultimately the result of mutation and selection. We therefore define the mutation rates solely in terms of mutation parameters.</p>
<div class="section" id="general-model">
<h4><a class="toc-backref" href="#id7">General model</a><a class="headerlink" href="#general-model" title="Permalink to this headline">¶</a></h4>
<p>There are four nucleotides, so in principle there are 12 mutation rates.
However in practice we expect to have the constraint:</p>
<div class="math" id="equation-Rmn">
<span class="eqno">(3)</span>\[R_{m \rightarrow n} = R_{m_c \rightarrow n_c}.\]</div>
<p>This constraint means, for example, that the rate from <em>A</em> to <em>G</em> is equal to the rate from <em>T</em> to <em>C</em>. The reason we expect this to be true is biological: a mutation from <em>m</em> to <em>n</em> on the sequenced strand causes a mutation of <span class="math">\(m_c\)</span> to <span class="math">\(n_c\)</span> on the complementary strand, so it is not possible to distinguish between these two types of mutations assuming that both strands are subject to the same mutational processes. This assumption reduces the number of independent mutation parameters to the following six parameters:</p>
<blockquote>
<div><ol class="arabic simple">
<li><span class="math">\(R_{A \rightarrow C}\)</span></li>
<li><span class="math">\(R_{A \rightarrow G}\)</span></li>
<li><span class="math">\(R_{A \rightarrow T}\)</span></li>
<li><span class="math">\(R_{C \rightarrow A}\)</span></li>
<li><span class="math">\(R_{C \rightarrow G}\)</span></li>
<li><span class="math">\(R_{C \rightarrow T}\)</span></li>
</ol>
</div></blockquote>
<p>Given these mutation, rates, it is useful to determine the expected evolutionary equilibrium codon frequencies in the case where all amino acids and stop codons are selectively neutral, and so evolution is driven entirely by the mutation process defined by Equations <a href="#equation-Rmn">(3)</a> and <a href="#equation-Qxy">(2)</a> (in other words, the expected evolutionary equilibrium codon frequencies in the case where <span class="math">\(F_{r,xy} = 1\)</span> in Equation <a href="#equation-Prxy">(1)</a>). In this case, the expected evolutionary equilibrium frequencies are determined entirely by the mutation rates as expected biologically. It is straightforward to show that for nonzero mutation rates that satisfy Equation <a href="#equation-Rmn">(3)</a>, the expected nucleotide frequency of a <em>A</em> or <em>T</em> nucleotide is <span class="math">\(\frac{1}{2}\times \frac{R_{C \rightarrow A} + R_{C \rightarrow T}}{R_{A \rightarrow C} + R_{A \rightarrow G} + R_{C \rightarrow A} + R_{C\rightarrow T}}\)</span> and the expected nucleotide frequency of a <em>C</em> or <em>G</em> nucleotide is <span class="math">\(\frac{1}{2} \times \frac{R_{A \rightarrow C} + R_{A \rightarrow G}}{R_{A \rightarrow C} + R_{A \rightarrow G} + R_{C \rightarrow A} + R_{C\rightarrow T}}\)</span>. The expected evolutionary equilibrium frequency <span class="math">\(q_x\)</span> of codon <em>x</em> in the case where all amino acids and stop codons are selectively neutral are simply the product of these nucleotide frequencies, so</p>
<div class="math" id="equation-qx">
<span class="eqno">(4)</span>\[q_x = \frac{1}{8} \times \frac{\left(R_{A \rightarrow C} + R_{A \rightarrow G} \right)^{\mathcal{N_{CG}}\left(x\right)}\times \left(R_{C \rightarrow A} + R_{C \rightarrow T} \right)^{\left(3 - \mathcal{N_{CG}}\left(x\right)\right)} }{\left(R_{A \rightarrow C} + R_{A \rightarrow G} + R_{C \rightarrow A} + R_{C\rightarrow T}\right)^3}\]</div>
<p>where <span class="math">\(\mathcal{N_{CG}}\left(x\right)\)</span> is the number of <em>C</em> or <em>G</em> nucleotides in codon <em>x</em>.</p>
<p>It is easy to verify that <span class="math">\(q_x\)</span> as defined by Equation <a href="#equation-qx">(4)</a> specifies the stationary state of a Markov process governed by the transition matrix defined by <span class="math">\(Q_{xy}\)</span> given in Equation <a href="#equation-Qxy">(2)</a>. For instance, if the diagonal elements of this matrix are defined as</p>
<div class="math" id="equation-Qxx">
<span class="eqno">(5)</span>\[Q_{xx} = 1 - \sum\limits_{y \ne x} Q_{xy}\]</div>
<p>then the matrix <span class="math">\(\left[Q_{xy}\right]\)</span> is a stochastic matrix, and so in general will have a unique stationary state with eigenvalue one. It is possible to verify by direct substitution that the <span class="math">\(q_x\)</span> defined by Equation <a href="#equation-qx">(4)</a> is that eigenvector (normalized to sum to one).</p>
</div>
<div class="section" id="reversible-model">
<h4><a class="toc-backref" href="#id8">Reversible model</a><a class="headerlink" href="#reversible-model" title="Permalink to this headline">¶</a></h4>
<p>Unfortunately, the process defined by <span class="math">\(Q_{xy}\)</span> via Equations <a href="#equation-Rmn">(3)</a> and <a href="#equation-Qxy">(2)</a> is not guaranteed to be <a class="reference external" href="http://en.wikipedia.org/wiki/Markov_chain#Reversible_Markov_chain">reversible</a>. The reversibility condition is</p>
<div class="math" id="equation-Qxy_reversible">
<span class="eqno">(6)</span>\[q_x Q_{xy} = q_y Q_{yx}.\]</div>
<p>To see that this condition is not satisfied, let <span class="math">\(x = AAT\)</span> and <span class="math">\(y = CAT\)</span> so that Equation <a href="#equation-Qxy_reversible">(6)</a> reduces</p>
<div class="math" id="equation-Rmn_reversible_algebra">
<span class="eqno">(7)</span>\[\begin{split}\left(R_{C \rightarrow A} + R_{C \rightarrow T} \right)^3 \times R_{A \rightarrow C} &amp;=&amp; \left(R_{A \rightarrow C} + R_{A \rightarrow G} \right) \times \left(R_{C \rightarrow A} + R_{C \rightarrow T} \right)^2 \times R_{C \rightarrow A} \implies \\
\left(R_{C \rightarrow A} + R_{C \rightarrow T} \right) \times R_{A \rightarrow C} &amp;=&amp; \left(R_{A \rightarrow C} + R_{A \rightarrow G} \right) \times R_{C \rightarrow A} \implies \\
R_{C \rightarrow T} \times R_{A \rightarrow C} &amp;=&amp; R_{A \rightarrow G} \times R_{C \rightarrow A}.\end{split}\]</div>
<p>In general this last condition is not guaranteed to be true. Therefore, to make the model <a class="reference external" href="http://en.wikipedia.org/wiki/Markov_chain#Reversible_Markov_chain">reversible</a>, it is necessary to define</p>
<div class="math" id="equation-Rmn_reversible">
<span class="eqno">(8)</span>\[R_{C \rightarrow T} = \frac{R_{A \rightarrow G} \times R_{C \rightarrow A}}{R_{A \rightarrow C}}.\]</div>
<p>The biological meaning of this constraint is not particularly clear or intuitive, but essentially it means that the path to getting from an <em>A</em> to a <em>T</em> via an intermediate mutation to <em>C</em> is equally likely to the path from getting from an <em>A</em> to a <em>T</em> via an intermediate mutation to <em>G</em>.</p>
<p>With the constraint in Equation <a href="#equation-Rmn_reversible">(8)</a> plus Equation <a href="#equation-Rmn">(3)</a>, there are a total of five unique mutation rates:</p>
<blockquote>
<div><ol class="arabic simple">
<li><span class="math">\(R_{A \rightarrow C}\)</span></li>
<li><span class="math">\(R_{A \rightarrow G}\)</span></li>
<li><span class="math">\(R_{A \rightarrow T}\)</span></li>
<li><span class="math">\(R_{C \rightarrow A}\)</span></li>
<li><span class="math">\(R_{C \rightarrow G}\)</span></li>
</ol>
</div></blockquote>
<p>The equilibrium codon frequencies with this constraint can be simplified to</p>
<div class="math" id="equation-qx_reversible">
<span class="eqno">(9)</span>\[q_x = \frac{1}{8} \times \frac{\left(R_{A \rightarrow C} + R_{A \rightarrow G} \right)^{\mathcal{N_{CG}}\left(x\right)}\times \left(R_{C \rightarrow A} + \frac{R_{A \rightarrow G} \times R_{C \rightarrow A}}{R_{A \rightarrow C}} \right)^{\left(3 - \mathcal{N_{CG}}\left(x\right)\right)} }{\left(R_{A \rightarrow C} + R_{A \rightarrow G} + R_{C \rightarrow A} + \frac{R_{A \rightarrow G} \times R_{C \rightarrow A}}{R_{A \rightarrow C}}\right)^3}.\]</div>
<p>Therefore, a reversible codon mutation model can be established in terms of five mutation rate parameters. Now Equation <a href="#equation-Qxy_reversible">(6)</a> is satisfied.</p>
<p>In many cases, only four of the five mutation rates are really free parameters (this would be the case unless specific numerical values are attached to the mutation rates, or unless it is not the case that the branch lengths are free parameters, or unless the sequences are date-stamped). In the case where the mutation rates are not scaled by any other means, we will use the convention of setting <span class="math">\(R_{A \rightarrow C} = 1\)</span> and then view the other four mutation rates as relative to this rate of one. In this case, it turns out (see <a class="reference external" href="http://dx.doi.org/10.1093/molbev/msu220">An experimentally informed evolutionary model improves phylogenetic fit to divergent lactamase homologs</a>) that</p>
<div class="math" id="equation-qx_simple">
<span class="eqno">(10)</span>\[q_x \propto \left(R_{C\rightarrow A}\right)^{\mathcal{N}_{AT}\left(x\right)}.\]</div>
<p>Essentially, Equation <a href="#equation-qx_simple">(10)</a> holds when the <em>mutationrates</em> parameter described in the <a class="reference internal" href="#input-file">Input file</a> has the value <em>freeparameters</em>.</p>
<p>When the <em>mutationrates</em> parameter described in <a class="reference internal" href="#input-file">Input file</a> does <strong>not</strong> have the value <em>freeparameters</em>, and we are instead specifying specific numerical rates, then it is no longer the case that <span class="math">\(R_{A \rightarrow C} = 1\)</span>. However, in this case we can still simplify the expression for <span class="math">\(q_x\)</span> to</p>
<div class="math" id="equation-qx_specificrates">
<span class="eqno">(11)</span>\[\begin{split}q_x &amp; \propto &amp; \left(R_{A \rightarrow C} + R_{A \rightarrow G} \right)^{\mathcal{N}_{CG}\left(x\right)}\times \left(R_{C \rightarrow A} + \frac{R_{A \rightarrow G} \times R_{C \rightarrow A}}{R_{A \rightarrow C}} \right)^{\left(3 - \mathcal{N}_{CG}\right)} \\
&amp; \propto &amp; \left(R_{A \rightarrow C} + R_{A \rightarrow G} \right)^{\mathcal{N}_{CG}\left(x\right)} \times \left(\frac{R_{C \rightarrow A}}{R_{A \rightarrow C}}\right)^{3 - \mathcal{N}_{CG}\left(x\right)} \times \left(R_{A \rightarrow C} + R_{A \rightarrow G} \right)^{3 - \mathcal{N}_{CG}\left(x\right)} \\
&amp; \propto &amp; \left(\frac{R_{C \rightarrow A}}{R_{A \rightarrow C}}\right)^{\mathcal{N}_{AT}\left(x\right)}.\end{split}\]</div>
</div>
</div>
<div class="section" id="amino-acid-preferences">
<h3><a class="toc-backref" href="#id9">Amino-acid preferences</a><a class="headerlink" href="#amino-acid-preferences" title="Permalink to this headline">¶</a></h3>
<p>For each codon site <em>r</em>, we are assumed to have measured the preference <span class="math">\(\pi_{r,a}\)</span> for each amino-acid <em>a</em>, where <span class="math">\(1 = \sum_a \pi_{r,a}\)</span>. These must be provided as input to the script based on experimental measurements.</p>
<p>The fixation probabilities are determined from these amino-acid preferences. The general idea is that mutations to higher preference amino acids are more likely to fix than mutations to lower preference amino acids. The exact quantitative relationships are described below.</p>
</div>
<div class="section" id="fixation-probabilities">
<h3><a class="toc-backref" href="#id10">Fixation probabilities</a><a class="headerlink" href="#fixation-probabilities" title="Permalink to this headline">¶</a></h3>
<p>We consider two different models for how the fixation probabilities <span class="math">\(F_{r,xy}\)</span> depend on the amino-acid preferences <span class="math">\(\pi_{r,xy}\)</span>. We refer to the first model as the <em>FracTolerated</em> model since it conceptually considers the preferences to be related to the fraction of genetic backgrounds that tolerate a given mutation. In this model,</p>
<div class="math" id="equation-Frxy_FracTolerated">
<span class="eqno">(12)</span>\[\begin{split}F_{r,xy} =
\begin{cases}
1 &amp; \mbox{if $\mathcal{A}\left(x\right) = \mathcal{A}\left(y\right)$} \\
\omega &amp; \mbox{if $\mathcal{A}\left(x\right) \ne \mathcal{A}\left(y\right)$ and $\pi_{r,\mathcal{A}\left(y\right)} \ge \pi_{r,\mathcal{A}\left(x\right)}$} \\
\omega \times \left(\frac{\pi_{r,\mathcal{A}\left(y\right)}}{\pi_{\mathcal{A}\left(x\right)}}\right)^{\beta} &amp; \mbox{otherwise.}
\end{cases}\end{split}\]</div>
<p>where <span class="math">\(\mathcal{A}\left(x\right)\)</span> denotes the amino acid encoded by codon <em>x</em> and the meaning of <span class="math">\(\omega\)</span> is discussed below.</p>
<p>We refer to the second model as the <em>HalpernBruno</em> model since it was originally introduced by <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/9656490">Halpern and Bruno, MBE, 1998</a>. This model interprets the amino-acid preferences <span class="math">\(\pi_{r,xy}\)</span> as being related to selection coefficients, and specifies,</p>
<div class="math" id="equation-Frxy_HalpernBruno">
<span class="eqno">(13)</span>\[\begin{split}F_{r,xy} =
\begin{cases}
1 &amp; \mbox{if $\mathcal{A}\left(x\right) = \mathcal{A}\left(y\right)$} \\
\omega &amp; \mbox{if $\mathcal{A}\left(x\right) \ne \mathcal{A}\left(y\right)$ and $\pi_{r,\mathcal{A}\left(x\right)} = \pi_{r,\mathcal{A}\left(y\right)}$} \\
\omega \times \frac{\beta \times \ln\left(\pi_{r,\mathcal{A}\left(y\right)} / \pi_{r,\mathcal{A}\left(x\right)}\right)}{1 - \left(\pi_{r,\mathcal{A}\left(x\right)} / \pi_{r,\mathcal{A}\left(y\right)}\right)^{\beta}} &amp; \mbox{otherwise.}
\end{cases}\end{split}\]</div>
<p>In both Equations <a href="#equation-Frxy_FracTolerated">(12)</a> and <a href="#equation-Frxy_HalpernBruno">(13)</a>, <span class="math">\(\omega\)</span> is a parameter that scales the rate of nonsynonymous versus synonymous mutations. A large value of <span class="math">\(\omega\)</span> means that nonsynonymous mutations are elevated relative to the expectation from the amino-acid preferences, and a small value means that they are depressed. Fixing <span class="math">\(\omega = 1\)</span> means that the amino-acid preferences naturally describe the relative rates of nonsynonymous and synonymous mutations by capturing all of the selection. This script allows models either with <span class="math">\(\omega = 1\)</span>, or <span class="math">\(\omega\)</span> as a free parameter.</p>
<p>In both Equations <a href="#equation-Frxy_FracTolerated">(12)</a> and <a href="#equation-Frxy_HalpernBruno">(13)</a>, <span class="math">\(\beta\)</span> is a parameter that corresponds to the relative strength of selection the experiment used to determind <span class="math">\(\pi_{r,a}\)</span> versus natural evolution. The actual preference in nature for an amino-acid is taken to be proportional to <span class="math">\(\pi_{r,a}^{\beta}\)</span>. Therefore, values of <span class="math">\(\beta &gt; 1\)</span> correspond to stronger selection in nature than in the experiments, and values of <span class="math">\(\beta &lt; 1\)</span> correspond to weaker selection in nature than in the experiments. This script allows models with either the fixed value of <span class="math">\(\beta = 1\)</span> or with <span class="math">\(\beta\)</span> as a free parameter.</p>
<p>Note that the definitions in both Equations <a href="#equation-Frxy_FracTolerated">(12)</a> and <a href="#equation-Frxy_HalpernBruno">(13)</a> are <a class="reference external" href="http://en.wikipedia.org/wiki/Markov_chain#Reversible_Markov_chain">reversible</a> with respect to <span class="math">\(\left(\pi_{r,x}\right)^{\beta}\)</span>, such that</p>
<div class="math" id="equation-Frxy_reversible">
<span class="eqno">(14)</span>\[\left(\pi_{r,x}\right)^{\beta} F_{r,xy} = \left(\pi_{r,y}\right)^{\beta} F_{r,yx}.\]</div>
</div>
<div class="section" id="equilibrium-frequencies-and-reversibility">
<h3><a class="toc-backref" href="#id11">Equilibrium frequencies and reversibility</a><a class="headerlink" href="#equilibrium-frequencies-and-reversibility" title="Permalink to this headline">¶</a></h3>
<p>Evaluation of the likelihood of a phylogenetic tree requires specifying expected evolutionary equilibrium frequency for the codons at the root node as well as the transition probabilities <span class="math">\(P_{r,xy}\)</span> given in Equation <a href="#equation-Prxy">(1)</a>. It is also mathematically convenient if the overall substitution model is <a class="reference external" href="http://en.wikipedia.org/wiki/Markov_chain#Reversible_Markov_chain">reversible</a>.</p>
<p>Here we show that for the choices of <span class="math">\(Q_{xy}\)</span> and <span class="math">\(F_{r,xy}\)</span> defined above, the substitution model is <a class="reference external" href="http://en.wikipedia.org/wiki/Markov_chain#Reversible_Markov_chain">reversible</a> and the equilibrium frequencies can be calculated.</p>
<p>Specifically, we let <span class="math">\(p_{r,x}\)</span> denote the evolutionary equilibrium frequency of codon <em>x</em> at site <em>r</em>. We assume stop codon variants are non-viable (they have <span class="math">\(\pi_{r,x} = 0\)</span>) and so the evolutionary equilibrium frequency for all stop codons is zero. This reduces us to 61 possible values for <em>x</em> at each site, corresponding to the the 61 non-stop codons.</p>
<p>The evolutionary equilibrium codon frequencies are given by:</p>
<div class="math" id="equation-prx">
<span class="eqno">(15)</span>\[\begin{split}p_{r,x} &amp;=&amp; \frac{\left(\pi_{r,\mathcal{A}\left(x\right)}\right)^{\beta} \times q_x }{\sum\limits_y \left(\pi_{r,\mathcal{A}\left(y\right)}\right)^{\beta} \times q_y} \\\end{split}\]</div>
<p>To see that these are the evolutionary equilibrium frequencies, define</p>
<div class="math" id="equation-Prxx">
<span class="eqno">(16)</span>\[P_{r,xx} = -\sum\limits_{y \ne x} P_{r,xy}.\]</div>
<p>With the definition in Equation <a href="#equation-Prxx">(16)</a>, the matrix <span class="math">\(\mathbf{I} + \mathbf{P_r}\)</span> is a <a class="reference external" href="http://en.wikipedia.org/wiki/Stochastic_matrix">stochastic matrix</a>, where <span class="math">\(\mathbf{I}\)</span> is the identity matrix and <span class="math">\(\mathbf{P_r} = \left[P_{r,xy}\right]\)</span>. For plausible values for the mutation rates and amino-acid preferences, <span class="math">\(\mathbf{I} + \mathbf{P_r}\)</span> will also be irreducible and acyclic. Therefore, it has a unique (within a scaling constant) principal eigenvector <span class="math">\(\mathbf{p_r} = \left[p_{r,x}\right]\)</span> with an eigenvalue of one that represents the equilibrium frequencies. In other words, <span class="math">\(\mathbf{p_r} = \mathbf{p_r} \left(\mathbf{I} + \mathbf{P_r}\right)\)</span>. It can be verified that Equation <a href="#equation-prx">(15)</a> defines such an eigenvector by writing this eigenvector equation in element-wise form. Immediately below, I verify this for the case where <span class="math">\(F_{r,xy}\)</span> is defined by Equation <a href="#equation-Frxy_FracTolerated">(12)</a>, and <span class="math">\(\pi_{r,\mathcal{A}\left(y\right)} &gt; \pi_{r,\mathcal{A}\left(x\right)}\)</span>. Other cases can be verified similarly.</p>
<div class="math" id="equation-Prx_eigenvector">
<span class="eqno">(17)</span>\[\begin{split}p_{r,x} &amp;=&amp; p_{r,x} + \sum\limits_y p_{r,y} P_{r,yx} \\
        &amp;=&amp; p_{r,x} + (\sum\limits_{y\ne x} p_{r,y} P_{r,yx}) + p_{r,x} P_{r,xx} \\
        &amp;=&amp; p_{r,x} + (\sum\limits_{y\ne x} p_{r,y} P_{r,yx}) - p_{r,x} \sum\limits_{y\ne x} P_{r,xy} \\
        &amp;=&amp; p_{r,x} + \sum\limits_{y \ne x} \left(p_{r,y} P_{r,yx} - p_{r,x} P_{r,xy}\right) \\
        &amp;=&amp; p_{r,x} + \sum\limits_{y \ne x} \left(p_{r,y} Q_{yx} F_{r,yx} - p_{r,x} Q_{xy} F_{r,xy}\right) \\
        &amp;=&amp; p_{r,x} + \sum\limits_{y \ne x} \left(\left(\pi_{r,\mathcal{A}\left(y\right)}\right)^{\beta} q_y Q_{yx} F_{r,yx} - \left(\pi_{r,\mathcal{A}\left(x\right)}\right)^{\beta} q_x Q_{xy} F_{r,xy}\right) \\
        &amp;=&amp; p_{r,x} + \sum\limits_{y \ne x} \left(\left(\pi_{r,\mathcal{A}\left(y\right)}\right)^{\beta} q_y Q_{yx} \left(\frac{\pi_{r,\mathcal{A}\left(x\right)}}{\pi_{r,\mathcal{A}\left(y\right)}}\right)^{\beta} - \left(\pi_{r,\mathcal{A}\left(x\right)}\right)^{\beta} q_x Q_{xy} \right) \\
        &amp;=&amp; p_{r,x} + \left(\pi_{r,\mathcal{A}\left(x\right)}\right)^{\beta} \sum\limits_{y \ne x} \left( q_y Q_{yx} - q_x Q_{xy} \right) \\
        &amp;=&amp; p_{r,x}\end{split}\]</div>
<p>where the last line follows from Equation <a href="#equation-Qxy_reversible">(6)</a>. This verifies that <span class="math">\(p_{r,x}\)</span> is the stationary state of the Markov process defined by <span class="math">\(P_{r,xy}\)</span>.</p>
<p>To verify that the substitution model is <a class="reference external" href="http://en.wikipedia.org/wiki/Markov_chain#Reversible_Markov_chain">reversible</a>, it is necessary to show that <span class="math">\(0 = p_{r,x} P_{r,xy} - p_{r,y} P_{r,yx}\)</span>. This is shown below for the case where <span class="math">\(F_{r,xy}\)</span> is defined by Equation <a href="#equation-Frxy_FracTolerated">(12)</a>, and <span class="math">\(\pi_{r,\mathcal{A}\left(y\right)} &gt; \pi_{r,\mathcal{A}\left(x\right)}\)</span>. Other cases can be verified similarly.</p>
<div class="math" id="equation-Prxy_reversible">
<span class="eqno">(18)</span>\[\begin{split}0 &amp;=&amp; p_{r,x} P_{r,xy} - p_{r,y} P_{r,yx} \\
  &amp;=&amp; \left(\pi_{r,\mathcal{A}\left(x\right)}\right)^{\beta} \times q_x  \times Q_{xy} \times F_{r,xy} - \left(\pi_{r,\mathcal{A}\left(y\right)}\right)^{\beta} \times q_y \times Q_{yx} \times F_{r,yx} \\
  &amp;=&amp; \left(\pi_{r,\mathcal{A}\left(x\right)}\right)^{\beta} \times q_x \times Q_{xy} - \left(\pi_{r,\mathcal{A}\left(y\right)}\right)^{\beta} \times q_y \times Q_{yx} \times \left(\frac{\pi_{r, \mathcal{A}\left(x\right)}}{\pi_{r, \mathcal{A}\left(y\right)}}\right)^{\beta} \\
  &amp;=&amp; q_x \times Q_{xy} - q_y \times Q_{yx} \\
  &amp;=&amp; 0\end{split}\]</div>
<p>where the last line follows from Equation <a href="#equation-Qxy_reversible">(6)</a>.
The fact that <span class="math">\(P_{r,xy}\)</span> defines a <a class="reference external" href="http://en.wikipedia.org/wiki/Markov_chain#Reversible_Markov_chain">reversible</a> Markov process with stationary state <span class="math">\(p_{r,x}\)</span> means that it is possible to define a symmetric matrix <span class="math">\(\mathbf{S_r}\)</span> such that</p>
<div class="math" id="equation-Sr">
<span class="eqno">(19)</span>\[\mathbf{S_r} \rm{diag}\left(\ldots, p_{r,x}, \ldots\right) = \mathbf{P_r}\]</div>
<p>where <span class="math">\(\rm{diag}\left(\ldots, p_{r,x}, \ldots\right)\)</span> is the diagonal matrix with <span class="math">\(p_{r,x}\)</span> along its diagonal. Noting <span class="math">\(\mathbf{S_r} = \mathbf{P_r} \rm{diag}\left(\ldots, \frac{1}{p_{r,x}}, \ldots\right)\)</span>, we have</p>
<div class="math" id="equation-Srxy">
<span class="eqno">(20)</span>\[\begin{split}S_{r,xy} =
\begin{cases}
\frac{P_{r,xy}}{p_{r,y}} = 0 &amp; \mbox{if $x$ and $y$ differ by more than one nucleotide mutation,} \\
\frac{P_{r,xy}}{p_{r,y}} = \left( \sum\limits_z \left(\pi_{r,\mathcal{A}\left(z\right)}\right)^{\beta} \times q_z \right) \frac{Q_{xy}}{q_y} \frac{F_{r,xy}}{\left(\pi_{r,\mathcal{A}\left(y\right)}\right)^{\beta}} &amp; \mbox{if $x$ and $y$ differ by one nucleotide,} \\
\frac{P_{r,xx}}{p_{r,x}} &amp; \mbox{otherwise.} \\
\end{cases}\end{split}\]</div>
<p>This matrix is symmetric since <span class="math">\(S_{r,xy} = S_{r,yx}\)</span> as can be verified from the fact that <span class="math">\(\frac{Q_{xy}}{q_y} = \frac{Q_{yx}}{q_x}\)</span> and <span class="math">\(\frac{F_{r,xy}}{\left(\pi_{r,\mathcal{A}\left(y\right)}\right)^{\beta}} = \frac{F_{r,yx}}{\left(\pi_{r,\mathcal{A}\left(x\right)}\right)^{\beta}}\)</span> as is guaranteed by the reversibility conditions in Equations <a href="#equation-Qxy_reversible">(6)</a> and <a href="#equation-Frxy_reversible">(14)</a>.</p>
</div>
</div>
<div class="section" id="requirements">
<h2><a class="toc-backref" href="#id12">Requirements</a><a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> : this script is part of the <a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> package.</li>
<li><a class="reference external" href="http://www.python.org/">Python</a> : this script has been tested with version 2.7.3.</li>
<li><a class="reference external" href="https://github.com/jbloom/mapmuts">mapmuts</a> : this script has been tested with version 1.0.</li>
<li><a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> : this script has been tested with version <tt class="docutils literal"><span class="pre">HYPHY</span> <span class="pre">2.220131214beta(MP)</span> <span class="pre">for</span> <span class="pre">Linux</span> <span class="pre">on</span> <span class="pre">x86_64</span></tt>.</li>
</ul>
</div>
<div class="section" id="running-the-script">
<h2><a class="toc-backref" href="#id13">Running the script</a><a class="headerlink" href="#running-the-script" title="Permalink to this headline">¶</a></h2>
<p>To run the script, create an <a class="reference internal" href="#input-file">Input file</a>  as described below. Then run the script from the command line using the input file as the single argument, as in:</p>
<div class="highlight-python"><div class="highlight"><pre>phyloExpCM_ExpModelOptimizeHyphyTree.py infile.txt
</pre></div>
</div>
<p>Note that running this script may take quite a while if <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> takes a long time to run. <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> can also start to consume a lot of memory for relatively big input trees &#8211; so you may need to use a computer with a large amount of RAM.</p>
<p><strong>If you are running multiple versions of this script then you should place them in separate subdirectories</strong>. This is important because otherwise the scripts may create temporary or permanent files with the same names that interfere with / overwrite each other.</p>
</div>
<div class="section" id="input-file">
<h2><a class="toc-backref" href="#id14">Input file</a><a class="headerlink" href="#input-file" title="Permalink to this headline">¶</a></h2>
<p>The input file specifies <cite>key</cite> / <cite>value</cite> pairs on individual lines in the format:</p>
<div class="highlight-python"><div class="highlight"><pre>key1 value1
key2 value2
</pre></div>
</div>
<p>Blank lines or lines
that begin with # are ignored (i.e. as comment lines).</p>
<p>The input file should contain the following keys:</p>
<ul>
<li><p class="first"><em>hyphypath</em> should contain a command that can be used to run <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> in batch mode. If you have installed <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> in the current search path, this might be something like <tt class="docutils literal"><span class="pre">HYPHYMP</span></tt> or <tt class="docutils literal"><span class="pre">HYPHYMP</span> <span class="pre">CPU=4</span></tt> (the latter command specifying how many CPUs to use for this multi-thread version) or <tt class="docutils literal"><span class="pre">HYPHYDEBUG</span></tt>. If <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> is not installed in the current search path, you will need to specify a full path to the executable.</p>
</li>
<li><p class="first"><em>fastafile</em> is the name of a FASTA file containing the sequences that comprise the tip nodes. These should be nucleotide coding sequences aligned at the codon level (not at the nucleotide level). Stop codons should be removed, and the aligned sequence length should be a multiple of three. Ambiguous nucleotide identities are not supported. The headers for the sequences should match the names of the tip nodes in <em>treefile</em>.</p>
</li>
<li><p class="first"><em>treefile</em> is a Newick-format phylogenetic tree giving the relationship among the sequences in <em>fastafile</em>. The names of the tip nodes must match those in <em>fastafile</em>. Branch lengths can be specified, but they will be optimized by <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a>. Branch supports can be present, but they are ignored and removed in the output tree.</p>
</li>
<li><p class="first"><em>aapreferences</em> is the name of a file giving the experimentally determined equilibrium amino-acid preferences. These might be the <span class="math">\(\pi_{r,a}\)</span> values inferred by the <a class="reference external" href="https://github.com/jbloom/mapmuts">mapmuts</a> script <tt class="docutils literal"><span class="pre">mapmuts_inferpreferences.py</span></tt> into the <tt class="docutils literal"><span class="pre">*_equilibriumpreferences.txt</span></tt> file. Specifically, <span class="math">\(\pi_{r,a}\)</span> represents the preference of site <em>r</em> for amino acid <em>a</em> under the constraint <span class="math">\(\sum_a \pi_{r,a} = 1\)</span>. This preference is the expected equilibrium amino-acid frequency of amino acid <em>a</em> at site <em>r</em> for a hypothetical evolving population where every amino-acid is equally likely to mutate to every other amino acid (so all frequencies would be 1/20 = 0.05 in the absence of selection).</p>
<p>The format of the file specified by <em>aapreferences</em> should match that of the <tt class="docutils literal"><span class="pre">*_equilibriumpreferences.txt</span></tt> file created by the <a class="reference external" href="https://github.com/jbloom/mapmuts">mapmuts</a> script <tt class="docutils literal"><span class="pre">mapmuts_inferpreferences.py</span></tt>. After an initial header line, each line gives the site number (1, 2, ... numbering) for all
specified sites, the wildtype amino acid(s) (not used by this script), the site entropy (not used by this script), and then the experimentally determined equilibrium amino-acid frequencies for each amino acid. These equilibrium frequencies can either include or exclude a stop codon (denoted by the * character) as a possible amino acid.
However, for the purposes of this script, stop codons are not considered possible
amino acid identities. So if there is value for a stop codon, it is set to zero and the remaining <span class="math">\(\pi_{r,a}\)</span> values are rescaled so that they sum to one. Here are a few example lines of the input file format:</p>
<div class="highlight-python"><div class="highlight"><pre>#SITE   WT_AA   SITE_ENTROPY    PI_A    PI_C    PI_D    PI_E    PI_F    PI_G    PI_H    PI_I    PI_K    PI_L    PI_M    PI_N    PI_P    PI_Q    PI_R    PI_S    PI_T    PI_V    PI_W    PI_Y    PI_*
1   M   2.85229 0.0113803   0.0402777   0.0153121   0.0320438   0.013312    0.00936795  0.026916    0.0192017   0.0224047   0.018926    0.554093    0.0445008   0.0138125   0.0212192   0.0131771   0.0152268   0.0237621   0.0102606   0.0369008   0.0367991   0.0211056
2   A   0.968359    0.878703    0.00384431  0.00390501  0.00815666  0.0048199   0.00244426  0.00333017  0.00258064  0.00391181  0.00094265  0.0248087   0.00238593  0.00064321  0.00288323  0.00610788  0.0012339   0.001455    0.0290568   0.0107545   0.00492073  0.00311186
3   S   2.93851 0.0295947   0.00304848  0.00227603  0.00668501  0.131168    0.000710502 0.0199653   0.00804841  0.000619715 0.366698    0.0132841   0.0223199   0.0048327   0.0170484   0.00875982  0.090712    0.0171888   0.0102176   0.177257    0.069395    0.000170873
</pre></div>
</div>
</li>
<li><p class="first"><em>mutationrates</em> specifies how we determine the mutation rate values.
As described in the <a class="reference internal" href="#mutation-rates">Mutation rates</a> section, with the the constraints in Equations <a href="#equation-Rmn">(3)</a> and <a href="#equation-Rmn_reversible">(8)</a>, there are a total of five unique mutation rate parameters:</p>
<blockquote>
<div><blockquote>
<div><ol class="arabic simple">
<li><span class="math">\(R_{A \rightarrow C}\)</span></li>
<li><span class="math">\(R_{A \rightarrow G}\)</span></li>
<li><span class="math">\(R_{A \rightarrow T}\)</span></li>
<li><span class="math">\(R_{C \rightarrow A}\)</span></li>
<li><span class="math">\(R_{C \rightarrow G}\)</span></li>
</ol>
</div></blockquote>
<p>There are two options for how these rates can be determined, specified by the choice of a value that you put here for the <em>mutationrates</em> key:</p>
<blockquote>
<div><ul>
<li><p class="first">The string <em>freeparameters</em>. In this case, five independent free parameters are created to describe the mutation rates, and are fitted by maximum likelihood. You should use this option if you do not have experimental data that determines the mutation rates ahead of time.</p>
</li>
<li><p class="first">The name of a file specifying experimentally measured mutation rates. Specifically, the entry for <em>A -&gt; G</em> represents the probability that a given site mutations from nucleotide <em>A</em> to nucleotide <em>G</em> in a unit of time given that it was already <em>A</em> (in other words, it is <span class="math">\(\Pr\left(G \rm{\; at\; time\;} t = 1 | A \rm{\; at\; time\;} t = 0\right)\)</span>. The five rates described above should be specified in the file as follows:</p>
<div class="highlight-python"><div class="highlight"><pre>AC 9.0e-6
AG 2.4e-5
AT 3.0e-6
CA 9.4e-6
CG 1.9e-6
</pre></div>
</div>
<p>The rates for other mutations are determined from these values as described by Equations <a href="#equation-Rmn">(3)</a> and <a href="#equation-Rmn_reversible">(8)</a>.</p>
</li>
</ul>
</div></blockquote>
</div></blockquote>
</li>
<li><p class="first"><em>scalefactor</em> is an option that has meaning only if <em>mutationrates</em> specifies the name of a file giving the mutation rates. If you instead set <em>mutationrates</em> to <em>freeparameters</em>, the <em>scalefactor</em> option is ignored if it is provided. The actual substitution model specified by Equation <a href="#equation-Prxy">(1)</a> involves multiplying the mutation rates by the fixation probabilities. So if you provide experimentally determined mutation rates and these are small numbers (as will be the case unless you rescale them yourself), then the entries in the substitution matrix will be very small. This could cause numerical underflow. Therefore, it is suggested that you set <em>scalefactor</em> to a number &gt; 0 that is sufficiently large that the product of <em>scalefactor</em> and the mutation rates is not dramatically smaller than one. For example, if the mutation rates are in the range of <span class="math">\(10^{-5}\)</span> to <span class="math">\(10^{-6}\)</span>, you might set <em>scalefactor</em> to 10000.0. Note that if there is perfect numerical accuracy, the value of <em>scalefactor</em> does not matter as long as it is &gt; 0.</p>
</li>
<li><p class="first"><em>fixationmodel</em> specifies how we convert the amino-acid preferences into fixation probabilities. It should be one of the following strings:</p>
<blockquote>
<div><ul class="simple">
<li><em>FracTolerated</em> specifies that we use the model given by Equation <a href="#equation-Frxy_FracTolerated">(12)</a>.</li>
<li><em>HalpernBruno</em> specifies that we use the model given by Equation <a href="#equation-Frxy_HalpernBruno">(13)</a>.</li>
</ul>
</div></blockquote>
</li>
<li><p class="first"><em>siteslist</em> is the name of a file that specifies all of the codon sites that are included in the analysis in 1, 2, ... numbering. Any lines beginning with a <em>#</em> character are skipped. The remaining lines should begin with an integer &gt;= 1 that gives a site number, followed by whitespace &#8211; anything following this whitespace is ignored. All of the numbers in this first column are taken to be the numbers of the codon sites included. You can use the <tt class="docutils literal"><span class="pre">*_equilibriumpreferences.txt</span></tt> file generated by the <a class="reference external" href="https://github.com/jbloom/mapmuts">mapmuts</a> script <tt class="docutils literal"><span class="pre">mapmuts_inferpreferences.py</span></tt> as the value for this argument, since it satisfies these formatting requirements. So potentially, you can set <em>siteslist</em> to the same file used for <em>aapreferences</em>. All sites listed must be present in the file specified by <em>aapreferences</em>.</p>
</li>
<li><p class="first"><em>keeptempfiles</em> specifies whether we create temporary files created by the script. A variety of files are created, primarily for the purpose of running <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> (i.e. many of these are <tt class="docutils literal"><span class="pre">*.ibf</span></tt> files for <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a>). These files are detailed in <a class="reference internal" href="#temporary-output-files">Temporary output files</a>. If <em>keeptempfiles</em> is <em>True</em>, then these files are retained after conclusion of the script. If <em>keeptempfiles</em> is <em>False</em>, then these files are deleted if the script terminates normally (they are retained if it ends in error). You might set <em>keeptempfiles</em> to <em>True</em> to create intermediate files to aid in debugging. You might set it to <em>False</em> if you want to avoid creating lots of output files.</p>
</li>
<li><p class="first"><em>outfileprefix</em> gives the name of the prefix pre-pended to the output files. You can also make it <em>None</em> if you don&#8217;t want to pre-pend any prefix.</p>
</li>
<li><p class="first"><em>fitomega</em> is an optional argument. If it is <em>None</em>, <em>False</em>, or is not specified, then the <span class="math">\(\omega\)</span> parameter defined in Equations <a href="#equation-Frxy_FracTolerated">(12)</a> and <a href="#equation-Frxy_HalpernBruno">(13)</a> is fixed to one. This corresponds to assuming that the amino-acid preferences capture all of the selection. If <em>fitomega</em> is instead specified with a value of <em>freeparameter</em>, then the value of <span class="math">\(\omega\)</span> is fit by maximum likelihood. This corresponds to the case where there is some additional selection against (or for) nonsynymous mutations beyond that captured by the amino-acid preferences.</p>
</li>
<li><p class="first"><em>fitbeta</em> is an optional argument. If it is <em>None</em>, <em>False</em>, or is not specified, then the <span class="math">\(\beta\)</span> parameter defined in Equations <a href="#equation-Frxy_FracTolerated">(12)</a> and <a href="#equation-Frxy_HalpernBruno">(13)</a> is fixed to one. This corresponds to equal strength of selection in the experiments and natural evolution. If <em>fitbeta</em> is instead given a value of <em>freeparameter</em>, then the value of <span class="math">\(\beta\)</span> is fit by maximum likelihood, thereby allowing for weaker (<span class="math">\(\beta &lt; 1\)</span>) or stronger (<span class="math">\(\beta &gt; 1\)</span>) selection in nature than in the experiments.</p>
</li>
<li><p class="first"><em>randomizepreferences</em> is an optional argument. If it is not specified or set to a value of <em>False</em>, nothing unusual is done. Otherwise it should be set to an integer &gt;= 1 that specifies the random number seed used to randomize the amino-acid preferences among sites. You might want to do this if you are testing that the model relies relies on accurate experimental determination of the amino-acid preferences at each site by scrambling these preferences.</p>
</li>
<li><p class="first"><em>persitelikelihoods</em> is an optional argument. If it is <em>None</em>, <em>False</em>, or not specified, the nothing is done. If you set it to <em>True</em>, then the script also reports the per-site likelihoods in a separate file. Specifically, the file with the suffix <tt class="docutils literal"><span class="pre">sitelikelihoods.txt</span></tt> (see <a class="reference internal" href="#output-files">Output files</a>) lists the per-site likelihood for each site after the mutation rates and branch lengths have been optimized by maximum likelihood. These parameters and branch lengths are <strong>not</strong> re-optimized for each site &#8211; rather, they are set to their global best values and then the likelihood contribution of each site is computed.</p>
</li>
</ul>
</div>
<div class="section" id="example-input-file">
<h2><a class="toc-backref" href="#id15">Example input file</a><a class="headerlink" href="#example-input-file" title="Permalink to this headline">¶</a></h2>
<p>Here is an example input file:</p>
<div class="highlight-python"><div class="highlight"><pre># Example input file for phyloExpCM_OptimizeDetectSelectionHyphy.py
hyphypath HYPHYMP CPU=4
fastafile Aligned_NPs.fasta
treefile codonphyml_tree.newick
aapreferences equilibriumpreferences.txt
mutationrates mutations.txt
scalefactor 10000.0
fixationmodel FracTolerated
siteslist siteslist.txt
keeptempfiles True
outfileprefix None
</pre></div>
</div>
</div>
<div class="section" id="output-files">
<h2><a class="toc-backref" href="#id16">Output files</a><a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>If you are running multiple instances of this script, <strong>it is very important that you run them in separate subdirectories.</strong> Otherwise they may create output files of the same name that overwrite or interfere with each other.</p>
<p>Any output files that already exist are overwritten.</p>
<p>Two types of output files are created. Primary output files contain the main results. Temporary output files are created to store intermediate results and operations (primarily related to running <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a>) &#8211; these files are deleted at the end of the script if <em>keeptempfiles</em> is <em>False</em>.</p>
<div class="section" id="primary-output-files">
<h3><a class="toc-backref" href="#id17">Primary output files</a><a class="headerlink" href="#primary-output-files" title="Permalink to this headline">¶</a></h3>
<p>The following primary output files are created. Each of these files begins with the prefix specified by <em>outfileprefix</em>, and then has the file name indicated.</p>
<blockquote>
<div><ul>
<li><p class="first"><tt class="docutils literal"><span class="pre">optimizedtree.newick</span></tt> is a Newick file containing the tree with the branch lengths optimized by <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> (using Equation <a href="#equation-Prxy">(1)</a>). The tree topology is unchanged from that specified in <em>newickfile</em>.</p>
</li>
<li><p class="first"><tt class="docutils literal"><span class="pre">optimizedtree_results.txt</span></tt> is a text file containing the results of optimizing the tree (this is the tree in <tt class="docutils literal"><span class="pre">optimizedtree.newick</span></tt>). This file gives the log likelihood, the number of branch lengths optimized, the number of free model parameters, and the values of any free parameters. If <em>mutationrates</em> specifies known mutation rates then there are no free parameters. If <em>mutationrates</em> is set to <em>freeparameters</em>, then there are five mutation rate parameters as described above (description of reversible model in <a class="reference internal" href="#mutation-rates">Mutation rates</a> section), but since one of these mutation rates is not independent of the branch lengths, there are really on four free parameters (since we have <span class="math">\(R_{A \rightarrow C} = 1\)</span> as described above). Here are example contents of this file:</p>
<div class="highlight-python"><div class="highlight"><pre>Log likelihood: -58.6512
Branch lengths optimized: 27
Model parameters optimized: 4
R_AG: 2.86282
R_AT: 0
R_CA: 1.62803
R_CG: 1.8586e-15
</pre></div>
</div>
<p>Here <em>R_AG</em> is the rate of mutation from <em>A</em> to <em>G</em> (<span class="math">\(R_{A \rightarrow G}\)</span>). If the mutation rates are not free parameters, then the last four lines giving these rates will be absent.</p>
<p>If <span class="math">\(\omega\)</span> is also being fit via setting <em>fitomega</em> to <em>freeparameter</em>, then the file also specifies the value of this parameter as in:</p>
<div class="highlight-python"><div class="highlight"><pre>omega 0.03132
</pre></div>
</div>
<p>If <span class="math">\(\beta\)</span> is also being fit via setting <em>fitbeta</em> to <em>freeparameter</em>, then the file also specifies the value of this parameter as in:</p>
<div class="highlight-python"><div class="highlight"><pre>beta 1.2314
</pre></div>
</div>
</li>
<li><p class="first"><tt class="docutils literal"><span class="pre">sitelikelihoods.txt</span></tt> is created if <em>persitelikelihoods</em> is <em>True</em>. In this case, the first line is a header line and all subsequent lines list the integer site number and the log likelihood with tabs between these two columns. The format is as follows:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c">#SITE   SITE_LOG_LIKELIHOODS</span>
<span class="mi">1</span>   <span class="o">-</span><span class="mf">4.792906363107919</span>
<span class="mi">2</span>   <span class="o">-</span><span class="mf">15.89808216031635</span>
<span class="mi">3</span>   <span class="o">-</span><span class="mf">17.19022748421175</span>
</pre></div>
</div>
</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="temporary-output-files">
<h3><a class="toc-backref" href="#id18">Temporary output files</a><a class="headerlink" href="#temporary-output-files" title="Permalink to this headline">¶</a></h3>
<p>The following temporary output files are created, largely for running <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a>. If <em>keeptempfiles</em> is <em>True</em>, then these files are kept in the directory. If <em>keeptempfiles</em> is <em>False</em>, then these files are deleted at the conclusion of the script if it terminates normally (if it ends in error, any created files are retained for debugging purposes). The temporary files are:</p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">Prxy.ibf</span></tt> is a <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> include batch file that specifies the substitution model in Equation <a href="#equation-Prxy">(1)</a>.</li>
<li><tt class="docutils literal"><span class="pre">coded_fastafile.fasta</span></tt> is a version of the sequence aligment in <em>fastafile</em> where the sequence headers have been changed to unique names that conform to <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> naming requirements.</li>
<li><tt class="docutils literal"><span class="pre">coded_treefile.newick</span></tt> is a version of the tree in <em>treefile</em> where the sequence names have been changed to match the renamed headers in <tt class="docutils literal"><span class="pre">coded_fastafile.fasta</span></tt>, so that they conform to <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> naming requirements.</li>
<li><tt class="docutils literal"><span class="pre">hyphy_optimizetree_cmds.bf</span></tt> is a <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> batch file that performs the operations to optimize the tree branch lengths and any global parameters (such as the mutation rates <span class="math">\(Q_{xy}\)</span>) over all sites.</li>
<li><tt class="docutils literal"><span class="pre">hyphy_optimizetree_output.txt</span></tt> is a file that contains the results of running <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> on <tt class="docutils literal"><span class="pre">hyphy_optimizetree_cmds.bf</span></tt>.</li>
<li><tt class="docutils literal"><span class="pre">coded_optimizedtree.newick</span></tt> is a version of the tree with <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> optimized branch lengths with the names still matching those that have been changed to <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> naming as in <tt class="docutils literal"><span class="pre">coded_fastafile.fasta</span></tt> and <tt class="docutils literal"><span class="pre">coded_treefile.newick</span></tt>. Except for the names being coded, this matches the primary output file with suffix <tt class="docutils literal"><span class="pre">optimizedtree.newick</span></tt>.</li>
<li>If <em>persitelikelihoods</em> is <em>True</em>, then for each site files of the name <tt class="docutils literal"><span class="pre">hyphy_site1_cmds.bf</span></tt> and <tt class="docutils literal"><span class="pre">hyphy_site1_output.txt</span></tt> are created for each site number (the example names above are for site <tt class="docutils literal"><span class="pre">1</span></tt>).</li>
</ul>
</div></blockquote>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">phyloExpCM_ExpModelOptimizeHyphyTree.py</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#substitution-models">Substitution models</a><ul>
<li><a class="reference internal" href="#substitution-model">Substitution model</a></li>
<li><a class="reference internal" href="#mutation-rates">Mutation rates</a><ul>
<li><a class="reference internal" href="#general-model">General model</a></li>
<li><a class="reference internal" href="#reversible-model">Reversible model</a></li>
</ul>
</li>
<li><a class="reference internal" href="#amino-acid-preferences">Amino-acid preferences</a></li>
<li><a class="reference internal" href="#fixation-probabilities">Fixation probabilities</a></li>
<li><a class="reference internal" href="#equilibrium-frequencies-and-reversibility">Equilibrium frequencies and reversibility</a></li>
</ul>
</li>
<li><a class="reference internal" href="#requirements">Requirements</a></li>
<li><a class="reference internal" href="#running-the-script">Running the script</a></li>
<li><a class="reference internal" href="#input-file">Input file</a></li>
<li><a class="reference internal" href="#example-input-file">Example input file</a></li>
<li><a class="reference internal" href="#output-files">Output files</a><ul>
<li><a class="reference internal" href="#primary-output-files">Primary output files</a></li>
<li><a class="reference internal" href="#temporary-output-files">Temporary output files</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="phyloExpCM_optimizeHyphyTree.html"
                        title="previous chapter">phyloExpCM_optimizeHyphyTree.py</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="phyloExpCM_FreqsFromAlignment.html"
                        title="next chapter">phyloExpCM_FreqsFromAlignment.py</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/phyloExpCM_ExpModelOptimizeHyphyTree.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="phyloExpCM_FreqsFromAlignment.html" title="phyloExpCM_FreqsFromAlignment.py"
             >next</a> |</li>
        <li class="right" >
          <a href="phyloExpCM_optimizeHyphyTree.html" title="phyloExpCM_optimizeHyphyTree.py"
             >previous</a> |</li>
        <li><a href="index.html">phyloExpCM 0.3 documentation</a> &raquo;</li>
          <li><a href="scripts.html" >Scripts</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2013, Jesse D. Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3a0.
    </div>
  </body>
</html>