<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.8.1: http://docutils.sourceforge.net/" />
<title>2014 analysis of beta-lactamase phylogenetic fit</title>
<script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
<style type="text/css">

/*
:Author: David Goodger (goodger@python.org)
:Id: $Id: html4css1.css 7056 2011-06-17 10:50:48Z milde $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.

See http://docutils.sf.net/docs/howto/html-stylesheets.html for how to
customize this style sheet.
*/

/* used to remove borders from tables and images */
.borderless, table.borderless td, table.borderless th {
  border: 0 }

table.borderless td, table.borderless th {
  /* Override padding for "table.docutils td" with "! important".
     The right padding separates the table cells. */
  padding: 0 0.5em 0 0 ! important }

.first {
  /* Override more specific margin styles with "! important". */
  margin-top: 0 ! important }

.last, .with-subtitle {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

a.toc-backref {
  text-decoration: none ;
  color: black }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

object[type="image/svg+xml"], object[type="application/x-shockwave-flash"] {
  overflow: hidden;
}

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em ;
  margin-right: 2em }

div.footer, div.header {
  clear: both;
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin: 0 0 0.5em 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.section-subtitle, h2.section-subtitle, h3.section-subtitle,
h4.section-subtitle, h5.section-subtitle, h6.section-subtitle {
  margin-top: 0.4em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

img.align-left, .figure.align-left, object.align-left {
  clear: left ;
  float: left ;
  margin-right: 1em }

img.align-right, .figure.align-right, object.align-right {
  clear: right ;
  float: right ;
  margin-left: 1em }

img.align-center, .figure.align-center, object.align-center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.align-left {
  text-align: left }

.align-center {
  clear: both ;
  text-align: center }

.align-right {
  text-align: right }

/* reset inner alignment in figures */
div.align-right {
  text-align: inherit }

/* div.align-center * { */
/*   text-align: left } */

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font: inherit }

pre.literal-block, pre.doctest-block, pre.math {
  margin-left: 2em ;
  margin-right: 2em }

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

span.section-subtitle {
  /* font-size relative to parent (h1..h6 element) */
  font-size: 80% }

table.citation {
  border-left: solid 1px gray;
  margin-left: 1px }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid 1px black;
  margin-left: 1px }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

table.docutils th.field-name, table.docinfo th.docinfo-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap ;
  padding-left: 0 }

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

ul.auto-toc {
  list-style-type: none }

</style>
</head>
<body>
<div class="document" id="analysis-of-beta-lactamase-phylogenetic-fit">
<h1 class="title">2014 analysis of beta-lactamase phylogenetic fit</h1>

<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#overview" id="id1">Overview</a></li>
<li><a class="reference internal" href="#sequence-numbering" id="id2">Sequence numbering</a></li>
<li><a class="reference internal" href="#software-used" id="id3">Software used</a></li>
<li><a class="reference internal" href="#input-files" id="id4">Input files</a></li>
<li><a class="reference internal" href="#running-the-analysis" id="id5">Running the analysis</a></li>
<li><a class="reference internal" href="#steps-in-the-analysis" id="id6">Steps in the analysis</a><ul>
<li><a class="reference internal" href="#extracting-amino-acid-preferences" id="id7">Extracting amino-acid preferences</a></li>
<li><a class="reference internal" href="#visualizing-the-amino-acid-preferences" id="id8">Visualizing the amino-acid preferences</a></li>
<li><a class="reference internal" href="#creation-of-sequence-alignment" id="id9">Creation of sequence alignment</a></li>
<li><a class="reference internal" href="#inspection-of-aligned-sequences" id="id10">Inspection of aligned sequences</a></li>
<li><a class="reference internal" href="#construction-of-phylogenetic-trees" id="id11">Construction of phylogenetic trees</a></li>
<li><a class="reference internal" href="#average-amino-acid-frequencies-in-the-alignments" id="id12">Average amino-acid frequencies in the alignments</a></li>
<li><a class="reference internal" href="#optimizing-the-trees-with-various-substitution-models" id="id13">Optimizing the trees with various substitution models</a></li>
<li><a class="reference internal" href="#comparison-of-per-site-likelihoods" id="id14">Comparison of per-site likelihoods</a></li>
</ul>
</li>
<li><a class="reference internal" href="#conclusions" id="id15">Conclusions</a></li>
</ul>
</div>
<div class="section" id="overview">
<h1><a class="toc-backref" href="#id1">Overview</a></h1>
<p>This is an analysis of the phylogenetic fit of an evolutionary model determined using deep mutational scanning of beta-lactamase. This is the analysis in the paper <a class="reference external" href="http://dx.doi.org/10.1093/molbev/msu220">An experimentally informed evolutionary model improves phylogenetic fit to divergent lactamase homologs</a>.</p>
<p>The results show that an evolutionary model using experimentally determined amino-acid preferences and mutation rates fit from the data outperforms standard phylogenetic evolutionary models.</p>
<p>The data comes from <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/24567513">Firnberg et al 2014</a>.</p>
<p>This analysis was performed by <a class="reference external" href="http://research.fhcrc.org/bloom/en.html">Jesse Bloom</a>.</p>
<p>The overall approach is to use the <a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> package to analyze a beta-lactamase sequence phylogeny (tree topology estimated by <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> and then fixed) with an experimentally determined evolutionary model based on data from <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/24567513">Firnberg et al 2014</a>, and a variety of standard codon substitution models.</p>
<p>Much of the analysis itself is performed using <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a>, which is run by <a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> using the scripts detailed in the <a class="reference external" href="http://jbloom.github.io/phyloExpCM">phyloExpCM documentation</a>.</p>
<p>The code, data, and results from the analysis are in the <tt class="docutils literal">./examples/2014Analysis_lactamase/</tt> subdirectory in the main <a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> package directory. This subdirectory contains all of the necessary input files (see <a class="reference internal" href="#input-files">Input files</a>) for running the analysis. It also contains a master script (<tt class="docutils literal">run_analysis.py</tt>) that can be used to regenerate the analysis with the command:</p>
<pre class="literal-block">
python run_analysis.py
</pre>
<p>provided that you have installed the software described in <a class="reference internal" href="#software-used">Software used</a>.</p>
</div>
<div class="section" id="sequence-numbering">
<h1><a class="toc-backref" href="#id2">Sequence numbering</a></h1>
<p>Beta-lactamase sequences are often numbered using the <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1151176/">Ambler</a> numbering scheme. This is the numbering scheme used by <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/24567513">Firnberg et al 2014</a>. However, the phylogenetic analysis described here converts all numbering to sequential 1, 2, ... numbering (beginning with the N-terminal methionine) of the wildtype lactamase parent used by <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/24567513">Firnberg et al 2014</a>. So this is sequential numbering of the protein encoded by the coding sequence in the file <tt class="docutils literal">TEM1_cds.fasta</tt> described in <a class="reference internal" href="#input-files">Input files</a>.</p>
<p>The file <tt class="docutils literal">sequential_to_Ambler.csv</tt> converts between the sequential and <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1151176/">Ambler</a> numbering scheme. This file is created from the data provided by <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/24567513">Firnberg et al 2014</a> as described in <a class="reference internal" href="#extracting-amino-acid-preferences">Extracting amino-acid preferences</a>.</p>
</div>
<div class="section" id="software-used">
<h1><a class="toc-backref" href="#id3">Software used</a></h1>
<p>The analysis was performed using the following software on the Fred Hutchinson computer cluster using the following Linux distribution:</p>
<pre class="literal-block">
Linux version 3.5.0-43-generic (buildd&#64;roseapple) (gcc version 4.6.3 (Ubuntu/Linaro 4.6.3-1ubuntu5) ) #66~precise1-Ubuntu SMP Thu Oct 24 14:52:23 UTC 2013
</pre>
<p>In order to replicate the analysis, you will need to use Linux or Mac OS X, and install the software listed below.</p>
<p>The specific versions of software used are specified. Other versions will likely work as well, but this has not been confirmed.</p>
<ul>
<li><p class="first"><a class="reference external" href="http://www.python.org/">Python</a> (version 2.7.3)</p>
</li>
<li><p class="first"><a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> (specifically, <a class="reference external" href="https://github.com/jbloom/phyloExpCM/tree/v0.3">phyloExpCM version 0.3</a>)</p>
</li>
<li><p class="first"><a class="reference external" href="http://www.numpy.org/">numpy</a> (version 1.6.1)</p>
</li>
<li><p class="first"><a class="reference external" href="http://matplotlib.org/">matplotlib</a> (version 1.3.1)</p>
</li>
<li><p class="first"><a class="reference external" href="http://biopython.org/wiki/Main_Page">Biopython</a> (version 1.58)</p>
</li>
<li><p class="first"><a class="reference external" href="https://github.com/jbloom/mapmuts">mapmuts</a> : (<a class="reference external" href="https://github.com/jbloom/mapmuts/tree/v1.0">mapmuts version 1.0</a>) was used to make the sequence logo plot and to do some basic sequence manipulation in <a class="reference external" href="http://www.python.org/">Python</a>.</p>
</li>
<li><p class="first"><a class="reference external" href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/needle.html">EMBOSS needle</a> (version 6.6.0). The executable with the name <tt class="docutils literal">needle</tt> is expected to be in the current search path.</p>
</li>
<li><p class="first"><a class="reference external" href="https://github.com/stamatak/standard-RAxML">RAxML</a> (version 7.7.6). The executable with the name <tt class="docutils literal"><span class="pre">raxmlHPC-SSE3</span></tt> is expected to be in the current search path. This is the SSE3 version, and was built using the command:</p>
<pre class="literal-block">
make -f Makefile.SSE3.gcc
</pre>
<p>and then copying the executable into the search path.</p>
</li>
<li><p class="first"><a class="reference external" href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a> (version 1.4.0) was used to open tree files for annotation, visualization, and saving of images.</p>
</li>
<li><p class="first"><a class="reference external" href="http://tree.bio.ed.ac.uk/software/pathogen/">Path-O-Gen</a> (version 1.4) was used to manually root the trees for visual display.</p>
</li>
<li><p class="first"><a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> (version 1.00 201306.18) was used to build phylogenetic trees. The executable with the name <tt class="docutils literal">codonphyml</tt> is expected to be in the current search path. This executable was compiled with the <em>omp</em> options to make it multi-threaded.</p>
</li>
<li><p class="first"><a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> (version HYPHY 2.220131214beta(MP) for Linux on x86_64) was used to optimize the branches and compute the likelihoods under the different models. The multiprocessor (<tt class="docutils literal">HYPHYMP</tt>) version of the executable was used. To install this, the following commands were run in the unzipped <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> source code directory after download:</p>
<pre class="literal-block">
cmake -DINSTALL_PREFIX=/home/jbloom/.local/ .
make MP2
make install
</pre>
</li>
<li><p class="first">This analysis was run on the FHCRC scientific computing cluster using <a class="reference external" href="https://computing.llnl.gov/linux/slurm/sbatch.html">sbatch</a> (version 2.6.2), and requires <a class="reference external" href="https://computing.llnl.gov/linux/slurm/sbatch.html">sbatch</a> to be installed to run as currently implemented. If you don't want to use <a class="reference external" href="https://computing.llnl.gov/linux/slurm/sbatch.html">sbatch</a>, all you need to do is modify the main <tt class="docutils literal">run_analysis.py</tt> script to not submit any jobs via <a class="reference external" href="https://computing.llnl.gov/linux/slurm/sbatch.html">sbatch</a>. The use of <a class="reference external" href="https://computing.llnl.gov/linux/slurm/sbatch.html">sbatch</a> in the current version simply allows more processors to be used simultaneously and so speeds up the analysis.</p>
</li>
<li><p class="first"><a class="reference external" href="http://www.imagemagick.org/script/convert.php">ImageMagick convert</a> 6.6.9-7 2012-08-17 was used to convert PDF files to JPGs.</p>
</li>
<li><p class="first"><a class="reference external" href="http://weblogo.threeplusone.com/">weblogo</a> (version 3.3) was used to make the sequence logo plot.</p>
</li>
<li><p class="first"><a class="reference external" href="http://pybrary.net/pyPdf/">PyPdf</a> (version 1.13) was used to make overlays on the sequence logo plot.</p>
</li>
</ul>
</div>
<div class="section" id="input-files">
<h1><a class="toc-backref" href="#id4">Input files</a></h1>
<p>Here are the input files that are necessary to repeat the analysis. These are the names assigned with the <tt class="docutils literal">./examples/2014Analysis_lactamase/</tt> subdirectory of the main <a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> package.</p>
<ul>
<li><p class="first"><tt class="docutils literal">run_analysis.py</tt> is a <a class="reference external" href="http://www.python.org/">Python</a> script that will perform the analysis assuming that you have all of the appropriate software (see <a class="reference internal" href="#software-used">Software used</a>) and input files. To run this script, use:</p>
<pre class="literal-block">
python run_analysis.py
</pre>
<p>at the command line.</p>
</li>
<li><p class="first"><tt class="docutils literal">get_preferences.py</tt> is a <a class="reference external" href="http://www.python.org/">Python</a> script that is used to extract the amino-acid preferences from <tt class="docutils literal">Firnberg_missense_mutation_fitnesses.csv</tt>.</p>
</li>
<li><p class="first"><tt class="docutils literal">get_treeseqs.py</tt> is a <a class="reference external" href="http://www.python.org/">Python</a> script that is used to construct the beta-lactamase sequence alignments.</p>
</li>
<li><p class="first"><tt class="docutils literal">TEM1_cds.fasta</tt> is a FASTA file giving the coding DNA sequence for the parent TEM1 beta-lactamase using by <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/24567513">Firnberg et al 2014</a> as reconstructed from <tt class="docutils literal">Data S2</tt> of their paper.</p>
</li>
<li><p class="first"><tt class="docutils literal">Firnberg_missense_mutation_fitnesses.csv</tt> is the Excel spreadsheet <tt class="docutils literal">Data S2</tt> from <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/24567513">Firnberg et al 2014</a> saved in CSV (comma separated values) format using Excel. This is the table that gives the estimated fitness of each amino-acid mutations, disregarding the possibility that different synonymous codons might have unequal fitnesses.</p>
</li>
<li><p class="first">Files containing the beta-lactamase sequences used to build the sequence alignments. The beta-lactamase sequences to use were taken from the <a class="reference external" href="http://www.lahey.org/Studies/">Lahey database</a>, which contains a curated listing of beta-lactamases. All <em>TEM</em> and <em>SHV</em> lactamases for which <a class="reference external" href="http://www.ncbi.nlm.nih.gov/genbank/">Genbank</a> nucleotide sequences were available were downloaded. First, the accession numbers for all sequences for which these were provided were parsed from the Excel tables in the <a class="reference external" href="http://www.lahey.org/Studies/">Lahey database</a> on March-9-2014. Then sequences were batch downloaded from <a class="reference external" href="http://www.ncbi.nlm.nih.gov/genbank/">Genbank</a> for all of these accessions for which such sequences were available. The accessions and the batch downloaded sequences are in the following files:</p>
<blockquote>
<ul class="simple">
<li><tt class="docutils literal">Lahey_TEM_GenbankAccessions.txt</tt></li>
<li><tt class="docutils literal">Lahey_TEM_GenbankSequences.gb</tt></li>
<li><tt class="docutils literal">Lahey_SHV_GenbankAccessions.txt</tt></li>
<li><tt class="docutils literal">Lahey_SHV_GenbankSequences.gb</tt></li>
</ul>
</blockquote>
</li>
<li><p class="first"><tt class="docutils literal">1XPB.pdb</tt> is PDB file <a class="reference external" href="http://www.rcsb.org/pdb/explore.do?structureId=1XPB">1XPB</a> of the crystal structure of TEM-1 beta-lactmase, as downloaded from the Protein Data Bank.</p>
</li>
<li><p class="first"><tt class="docutils literal">1XPB_renumbered.pdb</tt> is file <tt class="docutils literal">1XPB.pdb</tt> renumbered using <a class="reference external" href="http://dicsoft2.physics.iisc.ernet.in/pdbgoodies/inputpage.html">PDB Goodies</a> so that the numbering is sequential and matches the sequential numbering in <tt class="docutils literal">TEM1_cds.fasta</tt> in sequential numbering as described in <a class="reference internal" href="#sequence-numbering">Sequence numbering</a>.</p>
</li>
<li><p class="first"><tt class="docutils literal">1XPB_renumbered.dssp</tt> is the result of analyzing <tt class="docutils literal">1XPB_renumbered.pdb</tt> using the <a class="reference external" href="http://www.cmbi.ru.nl/hsspsoap/">DSSP webserver</a>. This contains the accessible surface area and secondary structure for the crystal structure.</p>
</li>
</ul>
</div>
<div class="section" id="running-the-analysis">
<h1><a class="toc-backref" href="#id5">Running the analysis</a></h1>
<p>The analysis in this directory can be repeated simply by running the master script <tt class="docutils literal">run_analysis.py</tt> from the command line with:</p>
<pre class="literal-block">
python run_analysis.py
</pre>
<p>(or run it using <a class="reference external" href="https://computing.llnl.gov/linux/slurm/sbatch.html">sbatch</a>).</p>
<p>Other than the files delineated under <a class="reference internal" href="#input-files">Input files</a>, all data files in this directory can be re-generated with this command. However, the analysis does involve a few manual steps to create nicely formatted and annotated tree images from the raw tree files produced by the script. These manual steps are described in the <a class="reference internal" href="#steps-in-the-analysis">Steps in the analysis</a> section below.</p>
<p>There is a variable called <em>use_existing_output</em> that is hardcoded into the beginning of this script. This variable is currently set to <em>True</em>, which means that if output already exists it is <strong>not</strong> overwritten and regenerated. If you want to regenerate the output (such as after changing input files or altering analysis parameters), then either manually delete the old output or change <em>use_existing_output</em> to <em>False</em> to force the script to overwrite old output. This script does <strong>not</strong> automatically test for dependencies among files, so if you change a file but keep <em>use_existing_output</em> to <em>True</em>, then you need to manually remove any old files with dependencies.</p>
<p>Note that <tt class="docutils literal">run_analysis.py</tt> uses multiple CPUs to speed up some of the analyses. If you don't have as many CPUs on your system, you might want to check that you are not overloading the system -- you can avoid this by using fewer CPUs, which will require you to manually modify the <tt class="docutils literal">run_analysis.py</tt> script.</p>
<p>Note that <tt class="docutils literal">run_analysis.py</tt> uses <a class="reference external" href="https://computing.llnl.gov/linux/slurm/sbatch.html">sbatch</a> by default to run some of its subsidiary programs -- if you don't want to do this, set the <em>use_sbatch</em> option to <em>False</em> within the program.</p>
</div>
<div class="section" id="steps-in-the-analysis">
<h1><a class="toc-backref" href="#id6">Steps in the analysis</a></h1>
<p>Here are the steps in the analysis. Except for a few indicated manual steps, all of this is done automatically by the <tt class="docutils literal">run_analysis.py</tt> script.</p>
<div class="section" id="extracting-amino-acid-preferences">
<h2><a class="toc-backref" href="#id7">Extracting amino-acid preferences</a></h2>
<p>The approach used here defines the evolutionary model in terms of the amino-acid preferences, where <span class="math">
\(\pi_{r,a}\)</span>
 denotes the preference of site <em>r</em> for amino acid <em>a</em>. <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/24567513">Firnberg et al 2014</a> provide their data in terms of the &quot;fitnesses&quot; of different mutations, which are computed from the deep sequencing counts. Some aspects of the procedure used to compute these &quot;fitnesses&quot; from the deep sequencing counts appear to be rather ad hoc. However, here we do not attempt to re-analyze the raw data of <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/24567513">Firnberg et al 2014</a> (which in any case is only partly provided in their publication), but instead simply take their fitness values at face value and convert them to something approximating the amino-acid preferences. This is done by the <a class="reference external" href="http://www.python.org/">Python</a> script <tt class="docutils literal">get_preferences.py</tt>, which can be run using the command:</p>
<pre class="literal-block">
python get_preferences.py
</pre>
<p>This script processes the data from <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/24567513">Firnberg et al 2014</a> in <tt class="docutils literal">Firnberg_missense_mutation_fitnesses.csv</tt> to generate the file <tt class="docutils literal">amino_acid_preferences.txt</tt>.</p>
<p>Specifically, the script works as follows:</p>
<ol class="arabic">
<li><p class="first">For each residue <em>r</em>, we read in the fitness value <span class="math">
\(w_{r,a}\)</span>
 for amino acid <em>a</em> in <tt class="docutils literal">Firnberg_missense_mutation_fitnesses.csv</tt> if such a value is present.</p>
</li>
<li><p class="first">For the wildtype amino acid at site <em>r</em>, we set <span class="math">
\(w_{r,a} = 1\)</span>
 rather than using the value (usually close to one) provided in <tt class="docutils literal">Firnberg_missense_mutation_fitnesses.csv</tt>. This is done because their method (see supplemental material of <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/24567513">Firnberg et al 2014</a>) defines the fitness of the wildtype to be one, and the values in the file that are usually close to but not exactly one are actually for synonymous mutations.</p>
</li>
<li><p class="first">Sometimes amino acids don't have a fitness specified. When this is the case, we take the mean of all fitness values at the site (including the wildtype amino acid values set to one in the previous step), and then assign this mean fitness to any amino acids with unknown fitness. This is probably non-ideal, because it seems likely that the missing amino acid fitnesses correspond to very low fitness identities. But because <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/24567513">Firnberg et al 2014</a> fail to provide raw sequencing data or computer code, it is unclear how to do any better.</p>
</li>
<li><p class="first">The preference for each amino acid is defined as proportional to its fitness, with the constraint that <span class="math">
\(\sum_a \pi_{r,a} = 1\)</span>
. So specifically,</p>
<div class="math">
\begin{equation*}
\pi_{r,a} = \frac{w_{r,a}}{\sum_{a'} w_{r,a'}}
\end{equation*}
</div>
<p>where <span class="math">
\(a'\)</span>
 ranges over all amino acids.</p>
</li>
<li><p class="first">The preferences computed using the aforementioned formula are written to the output file <tt class="docutils literal">amino_acid_preferences.txt</tt>.
Note that the residues in the created <tt class="docutils literal">amino_acid_preferences.txt</tt> file are numbered sequentially rather than using the <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1151176/">Ambler</a> numbering (see <a class="reference internal" href="#sequence-numbering">Sequence numbering</a>).</p>
</li>
<li><p class="first">The <tt class="docutils literal">get_preferences.py</tt> script also creates the file <tt class="docutils literal">sequential_to_Ambler.csv</tt>, which converts between the sequential and <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1151176/">Ambler</a> numbering schemes (see <a class="reference internal" href="#sequence-numbering">Sequence numbering</a>).</p>
</li>
</ol>
</div>
<div class="section" id="visualizing-the-amino-acid-preferences">
<h2><a class="toc-backref" href="#id8">Visualizing the amino-acid preferences</a></h2>
<p>The amino-acid preferences are visualized using a logo plot created by the <a class="reference external" href="https://github.com/jbloom/mapmuts">mapmuts</a> script <tt class="docutils literal">mapmuts_siteprofileplots.py</tt>, which is detailed in the <a class="reference external" href="http://jbloom.github.io/mapmuts/">mapmuts documentation</a>. Briefly, this script uses <a class="reference external" href="http://weblogo.threeplusone.com/">weblogo</a> to visualize the preferences by making a plot in which the heights of amino acids are equal to their preferences at each site. Secondary structure and relative solvent accessibility information calculated using the <a class="reference external" href="http://www.cmbi.ru.nl/hsspsoap/">DSSP webserver</a> from PDB <a class="reference external" href="http://www.rcsb.org/pdb/explore.do?structureId=1XPB">1XPB</a> is overlaid on the plot. The residues in this plot are numbered using the <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1151176/">Ambler</a> numbering scheme (see <a class="reference internal" href="#sequence-numbering">Sequence numbering</a>).</p>
<p>The main <tt class="docutils literal">run_analysis.py</tt> script runs <tt class="docutils literal">mapmuts_siteprofileplots.py</tt> after creating its input file <tt class="docutils literal">mapmuts_siteprofileplots_infile.txt</tt>. The resulting image is in <tt class="docutils literal">lactamase_site_preferences_logoplot.pdf</tt>. A JPG version of this plot (<tt class="docutils literal">lactamase_site_preferences_logoplot.jpg</tt>) is also created using <a class="reference external" href="http://www.imagemagick.org/script/convert.php">ImageMagick convert</a>. The JPG is shown below:</p>
<div class="figure align-center">
<img alt="lactamase_site_preferences_logoplot.jpg" src="lactamase_site_preferences_logoplot.jpg" style="width: 80%;" />
<p class="caption">Visual display of amino-acid preferences. Letter heights are proportional to the preference for that amino acid at that site. Secondary structure (SS) and relative solvent accessibility information is overlaid on the plots. Amino-acid letters are colored according to a hydrophobicity scale.</p>
</div>
</div>
<div class="section" id="creation-of-sequence-alignment">
<h2><a class="toc-backref" href="#id9">Creation of sequence alignment</a></h2>
<p>The creation of the sequence alignment <tt class="docutils literal">aligned_lactamases.fasta</tt> is performed by the <a class="reference external" href="http://www.python.org/">Python</a> script <tt class="docutils literal">get_treeseqs.py</tt> using the command:</p>
<pre class="literal-block">
python get_treeseqs.py
</pre>
<p>The steps performed by this script are as follows:</p>
<ol class="arabic simple">
<li>The CDS sequences are extracted from the Genbank sequences in <tt class="docutils literal">Lahey_TEM_GenbankSequences.gb</tt> and <tt class="docutils literal">Lahey_SHV_GenbankSequences.gb</tt> for all records with a single specified CDS that encodes a translatable protein and has not ambiguous nucleotide identities.</li>
<li>The CDS sequences are pairwise aligned with the reference sequence in <tt class="docutils literal">TEM1_cds.fasta</tt> constructing codon alignments from the protein alignments, and any gaps relative to the reference sequence are stripped away.</li>
<li>Any sequences that do have less than 60% sequence identity to <tt class="docutils literal">TEM1_cds.fasta</tt> among alignable sites or have more than 20% gaps are removed.</li>
<li>Aligned sequences are examined for the number of nucleotide differences with other sequences. When there are multiple sequences that have less than four differences from another sequence, only one representative is retained. This reduces the number of highly similar sequences.</li>
<li>The remaining set of unique aligned sequences is written to <tt class="docutils literal">aligned_lactamases.fasta</tt>.</li>
</ol>
</div>
<div class="section" id="inspection-of-aligned-sequences">
<h2><a class="toc-backref" href="#id10">Inspection of aligned sequences</a></h2>
<p>The aligned sequences in <tt class="docutils literal">aligned_lactamases.fasta</tt> are subjected to the following manual steps to check the quality of the data:</p>
<ol class="arabic">
<li><p class="first">The <tt class="docutils literal">aligned_lactamases.fasta</tt> file was manually checked for recombinant sequences using the online interface to <a class="reference external" href="http://www.datamonkey.org/">DataMonkey</a> to run the <em>SBP</em> and <em>GARD</em> programs. Briefly:</p>
<blockquote>
<ul class="simple">
<li>Using <em>SBP</em>, there was no evidence of recombination using the recommended criterion of <em>cAIC</em>.</li>
<li>Using <em>GARD</em>, there was no evidence of recombination.</li>
</ul>
</blockquote>
</li>
<li><p class="first">A quick phylogenetic tree was built using <a class="reference external" href="https://github.com/stamatak/standard-RAxML">RAxML</a> to make sure that the tree topology seemed reasonable. The <a class="reference external" href="https://github.com/stamatak/standard-RAxML">RAxML</a> output was placed in the subdirectory <tt class="docutils literal">./RAxML_output/</tt>. The following command was executed:</p>
<pre class="literal-block">
raxmlHPC-SSE3 -w ./RAxML_output/ -n aligned_lactamases -p 1 -m GTRCAT -s aligned_lactamases.fasta
</pre>
<p>Note that in the actual command, the full path to <tt class="docutils literal">./RAxML_output/</tt> is specified. This created the <a class="reference external" href="https://github.com/stamatak/standard-RAxML">RAxML</a> tree as <tt class="docutils literal">/RAxML_output/RAxML_bestTree.aligned_lactamases</tt>, as well as several other files in <tt class="docutils literal">./RAxML_output/</tt>. The tree was manually visualized with <a class="reference external" href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a> to confirm that it separated into the two expected clades of TEM and SHV sequences.</p>
</li>
</ol>
</div>
<div class="section" id="construction-of-phylogenetic-trees">
<h2><a class="toc-backref" href="#id11">Construction of phylogenetic trees</a></h2>
<p>High-quality phylogenetic trees were then constructed using <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a>, which is able to build maximum-likelihood trees with codon substitution models.</p>
<p>The <a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> script <tt class="docutils literal">phyloExpCM_runcodonPhyML.py</tt> (see <a class="reference external" href="http://jbloom.github.io/phyloExpCM">phyloExpCM documentation</a> for details) was used to run <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a>. Trees were built using two different substitution models, with each tree in its own subdirectory within <tt class="docutils literal">./examples/2014Analysis_lactamase/</tt>. The substitution models used to build the trees:</p>
<blockquote>
<ul>
<li><p class="first">The <em>GY94</em> codon model (<a class="reference external" href="http://mbe.oxfordjournals.org/content/11/5/725.full.pdf">Goldman and Yang 1994</a>) with:</p>
<blockquote>
<ul class="simple">
<li>The equilibrium codon frequencies estimated empirically using the <a class="reference external" href="http://www.plosone.org/article/info%3Adoi/10.1371/journal.pone.0011230">CF3x4</a> method.</li>
<li>A single transition / transversion ratio (<em>kappa</em>) estimated by maximum likelihood.</li>
<li>The dN/dS ratio (<em>omega</em>) drawn from four discrete gamma-distributed categories (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/7932792">Yang 1994</a>) with the shape parameter and mean estimated by maximum likelihood.</li>
</ul>
</blockquote>
<p>The tree construction was performed in the <tt class="docutils literal">./GY94_codonPhyML_tree/</tt> subdirectory. The tree itself is in the file <tt class="docutils literal">./GY94_codonPhyML_tree/codonphyml_tree.newick</tt>.</p>
</li>
<li><p class="first">The <em>KOSI07</em> empirical codon model (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/17400572">Kosiol et al 2007</a>), using the variant denoted as <em>ECM+F+omega+1kappa(tv)</em> in <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/17400572">Kosiol et al 2007</a>. Specifically:</p>
<blockquote>
<ul class="simple">
<li>The equilibrium codon frequencies were estimated empirically using the <em>F</em> method (empirical estimation of 60 frequencies for the 61 non-stop codons).</li>
<li>The relative decrease in transversions versus transitions, <em>kappa(tv)</em>, estimated by maximum likelihood.</li>
<li>The relative elevation in nonsynonymous over synonymous mutations (<em>omega</em>) drawn from four gamma-distributed categories with shape parameter and mean estimated by maximum likelihood.</li>
</ul>
</blockquote>
<p>The tree construction was performed in the <tt class="docutils literal">./KOSI07_codonPhyML_tree/</tt> subdirectory. The tree itself is in the file <tt class="docutils literal">./KOSI07_codonPhyML_tree/codonphyml_tree.newick</tt>.</p>
</li>
</ul>
</blockquote>
<p>To visualize these trees, the following <strong>manual</strong> steps were performed for the tree in each subdirectory:</p>
<blockquote>
<ol class="arabic">
<li><p class="first">The <tt class="docutils literal">codonphyml_tree.newick</tt> file was re-rooted using <a class="reference external" href="http://tree.bio.ed.ac.uk/software/pathogen/">Path-O-Gen</a> and then opened in <a class="reference external" href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a>, adjusted for attractive visual display, and saved to the file <tt class="docutils literal">annotated_tree.trees</tt>.</p>
</li>
<li><p class="first">A PDF image of the tree in the <tt class="docutils literal">annotated_tree.trees</tt> file was saved manually with <a class="reference external" href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a>, and then converted to a JPG with <a class="reference external" href="http://www.imagemagick.org/script/convert.php">ImageMagick convert</a> using the command:</p>
<pre class="literal-block">
convert -density 300 annotated_tree.pdf annotated_tree.jpg
</pre>
<p>The tree images created by these steps are shown below:</p>
<blockquote>
<div class="figure align-center">
<img alt="GY94_codonPhyML_tree/annotated_tree.jpg" src="GY94_codonPhyML_tree/annotated_tree.jpg" style="width: 65%;" />
<p class="caption">Image of the tree generated by <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> for the <em>GY94</em> substitution model (file <tt class="docutils literal">./GY94_codonPhyML_tree/annotated_tree.jpg</tt>). The <em>TEM</em> sequences are colored red, while the <em>SHV</em> sequences are colored blue.</p>
</div>
<div class="figure align-center">
<img alt="KOSI07_codonPhyML_tree/annotated_tree.jpg" src="KOSI07_codonPhyML_tree/annotated_tree.jpg" style="width: 65%;" />
<p class="caption">Image of the tree generated by <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> for the <em>KOSI07</em> substitution model (file <tt class="docutils literal">./KOSI07_codonPhyML_tree/annotated_tree.jpg</tt>). The <em>TEM</em> sequences are colored red, while the <em>SHV</em> sequences are colored blue.</p>
</div>
</blockquote>
</li>
</ol>
</blockquote>
<p>The <a class="reference external" href="http://en.wikipedia.org/wiki/Robinson%E2%80%93Foulds_metric">Robinson-Foulds distance</a> between the trees computed with the <em>GY94</em> and <em>KOSI07</em> models was computed with <a class="reference external" href="https://github.com/stamatak/standard-RAxML">RAxML</a>, and the results written to the directory <tt class="docutils literal">./RobinsonFouldsDistances/</tt>. This distance provides a quantitative measure of the similarity between the tree topologies.</p>
<p>Each of these trees was also parsed into the subtrees representing the <em>TEM</em> and the <em>SHV</em> sequences keeping the tree topology estimated jointly. The subtree parsing is done automatically by the <tt class="docutils literal">run_analysis.py</tt> script using <a class="reference external" href="http://biopython.org/wiki/Main_Page">BioPython</a>. These subtrees have the names:</p>
<blockquote>
<ul class="simple">
<li><tt class="docutils literal">./KOSI07_codonPhyML_tree/codonphyml_tree_SHV.newick</tt> : the <em>SHV</em> subtree from <tt class="docutils literal">./KOSI07_codonPhyML_tree/codonphyml_tree.newick</tt></li>
<li><tt class="docutils literal">./KOSI07_codonPhyML_tree/codonphyml_tree_TEM.newick</tt></li>
<li><tt class="docutils literal">./GY94_codonPhyML_tree/codonphyml_tree_SHV.newick</tt></li>
<li><tt class="docutils literal">./GY94_codonPhyML_tree/codonphyml_tree_TEM.newick</tt></li>
</ul>
</blockquote>
<p>The sequence alignments for these subtrees are in the files</p>
<blockquote>
<ul class="simple">
<li><tt class="docutils literal">aligned_GY94_SHV.fasta</tt></li>
<li><tt class="docutils literal">aligned_GY94_TEM.fasta</tt></li>
<li><tt class="docutils literal">aligned_KOSI07_SHV.fasta</tt></li>
<li><tt class="docutils literal">aligned_KOSI07_TEM.fasta</tt></li>
</ul>
</blockquote>
</div>
<div class="section" id="average-amino-acid-frequencies-in-the-alignments">
<h2><a class="toc-backref" href="#id12">Average amino-acid frequencies in the alignments</a></h2>
<p>The overall average amino-acid frequencies for each alignment (averaged over all sites and sequences) were computed using the <a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> script <tt class="docutils literal">phyloExpCM_FreqsFromAlignment.py</tt>.</p>
<p>Within the <tt class="docutils literal">./examples/2014Analysis_lactamase/</tt> these files have the following names:</p>
<blockquote>
<ul class="simple">
<li><tt class="docutils literal">avg_aa_freqs_aligned_lactamases.txt</tt> : the averages over all of the lactamase sequences in <tt class="docutils literal">aligned_lactamases.fasta</tt></li>
<li><tt class="docutils literal">avg_aa_freqs_aligned_GY94_TEM.txt</tt> : the averages over just the sequences in <tt class="docutils literal">aligned_GY94_TEM.fasta</tt></li>
<li><tt class="docutils literal">avg_aa_freqs_aligned_KOSI07_TEM.txt</tt></li>
<li><tt class="docutils literal">avg_aa_freqs_aligned_GY94_SHV.txt</tt></li>
<li><tt class="docutils literal">avg_aa_freqs_aligned_KOSI07_SHV.txt</tt></li>
</ul>
</blockquote>
<p>Although these files list a line of frequencies for each site, the frequencies reported for each site are the <strong>average</strong> over all sites and sequences. These frequencies are used in the models that have the same mutation structure as the experimentally informed models, but replace the preferences with the average amino-acid frequencies. These frequencies are calculated by adding a pseudocount of one to each amino-acid type (a very small correction as there are many sequences and sites).</p>
</div>
<div class="section" id="optimizing-the-trees-with-various-substitution-models">
<h2><a class="toc-backref" href="#id13">Optimizing the trees with various substitution models</a></h2>
<p>This is the heart of the analysis: to use various codon substitution models to optimize the branch lengths and likelihood for the tree topologies estimated with <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> and the <em>GY94</em> or <em>KOSI07</em> substitution models as described in <a class="reference internal" href="#construction-of-phylogenetic-trees">Construction of phylogenetic trees</a>. The analyses are also done for the <em>TEM</em> and <em>SHV</em> subtrees. This analysis compares the experimentally determined codon substitution models with various other substitution models.</p>
<p>This analysis is done with the <a class="reference external" href="https://github.com/jbloom/phyloExpCM">phyloExpCM</a> scripts <tt class="docutils literal">phyloExpCM_optimizeHyphyTree.py</tt> (for standard substitution models) and <tt class="docutils literal">phyloExpCM_ExpModelOptimizeHyphyTree.py</tt> (for substitution models using the experimentally determined amino-acid preferences), both of which are described in the <a class="reference external" href="http://jbloom.github.io/phyloExpCM">phyloExpCM documentation</a>. These scripts uses <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> to optimize the branch lengths and any substitution model free parameters while keeping the tree topologies fixed.</p>
<p>The results of the optimizations for the various substitution models are in the subdirectory <tt class="docutils literal">./codonmodel_optimized_trees/</tt>. Within this subdirectory are further subdirectories containing the results of optimizing the branch lengths with <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> using the indicated codon substitution models. For example, <tt class="docutils literal"><span class="pre">./codonmodel_optimized_trees/Tree-KOSI07_Model-KOSI07_F_omega-global-gamma4_rates-one/</span></tt> contains one such model. Within this subdirectory, the files of interest are:</p>
<blockquote>
<ul class="simple">
<li>The tree with branch lengths optimized by <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> using the substitution model in question is in a file called either <tt class="docutils literal">hyphy_tree.newick</tt> or <tt class="docutils literal">optimizedtree.newick</tt>.</li>
<li>A summary of the <a class="reference external" href="http://www.hyphy.org/w/index.php/Main_Page">HYPHY</a> results including the log likelihood is found in a file called either <tt class="docutils literal">hyphy_output.txt</tt> or <tt class="docutils literal">optimizedtree_results.txt</tt>.</li>
<li><tt class="docutils literal">*_infile.txt</tt> the input file used to run either <tt class="docutils literal">phyloExpCM_optimizeHyphyTree.py</tt> or <tt class="docutils literal">phyloExpCM_ExpModelOptimizeHyphyTree.py</tt>.</li>
</ul>
</blockquote>
<p>The models are:</p>
<blockquote>
<ul>
<li><p class="first"><em>HalpernBruno</em> : the model described as <em>HalpernBruno</em> in the <a class="reference external" href="http://jbloom.github.io/phyloExpCM">phyloExpCM documentation</a> for <tt class="docutils literal">phyloExpCM_OptimizeDetectSelection.py</tt>. Briefly, the fixation probabilities <span class="math">
\(F_{r,xy}\)</span>
 are determined from the amino-acid preferences <span class="math">
\(\pi_{r,a}\)</span>
 using formula of <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/9656490">Halpern and Bruno 1998</a> as</p>
<blockquote>
<div class="math">
\begin{equation*}
F_{r,xy} =
\begin{cases}
1 &amp; \mbox{if $\pi_{r,\mathcal{A}\left(x\right)} = \pi_{r,\mathcal{A}\left(y\right)}$} \\
\frac{\beta \times \ln\left(\pi_{r,\mathcal{A}\left(y\right)} / \pi_{r,\mathcal{A}\left(x\right)}\right)}{1 - \left(\pi_{r,\mathcal{A}\left(x\right)} / \pi_{r,\mathcal{A}\left(y\right)}\right)^{\beta}} &amp; \mbox{otherwise.}
\end{cases}
\end{equation*}
</div>
</blockquote>
<p>The mutation rates involve the fitting of four free parameters as described in the <a class="reference external" href="http://jbloom.github.io/phyloExpCM">phyloExpCM documentation</a> for <tt class="docutils literal">phyloExpCM_OptimizeDetectSelection.py</tt>. The stringency parameter <span class="math">
\(\beta\)</span>
 is either fixed to one or optimized.</p>
</li>
<li><p class="first"><em>FracTolerated</em> : the model described as <em>FracTolerated</em> in the <a class="reference external" href="http://jbloom.github.io/phyloExpCM">phyloExpCM documentation</a> for <tt class="docutils literal">phyloExpCM_OptimizeDetectSelection.py</tt>. This model differs from <em>HalpernBruno</em> in that the fixation probabilities are now determined as</p>
<blockquote>
<div class="math">
\begin{equation*}
F_{r,xy} =
\begin{cases}
1 &amp; \mbox{if $\pi_{r,\mathcal{A}\left(y\right)} \ge \pi_{r,\mathcal{A}\left(x\right)}$} \\
\left(\frac{\pi_{r,\mathcal{A}\left(y\right)}}{\pi_{\mathcal{A}\left(x\right)}}\right)^{\beta} &amp; \mbox{otherwise.}
\end{cases}
\end{equation*}
</div>
</blockquote>
</li>
<li><p class="first"><em>HalpernBrunorandomized</em> : like <em>HalpernBruno</em> but with the experimentally determined amino-acid preferences randomized among sites. This model is not expected to be good because there is no longer any accurate site-specific information.</p>
</li>
<li><p class="first"><em>FracToleratedrandomized</em> : like <em>FracTolerated</em> but with the experimentally determined amino-acid preferences randomized among sites.</p>
</li>
<li><p class="first"><em>avgaafreqsHalpernBruno</em> : this model uses the mutation model and fixation model of <em>HalpernBruno</em>, but does not use the experimentally determined amino-acid preferences. Instead, the amino-acid preferences for all sites are simply set based on the amino-acid frequencies obtained from averaging all sites and sequences in the alignment (see <a class="reference internal" href="#average-amino-acid-frequencies-in-the-alignments">Average amino-acid frequencies in the alignments</a>). So this is a null model that captures the mutation process of the <em>HalpernBruno</em> model but isn't site specific. It does contain 60 parameters (the amino-acid frequencies) that are estimated empirically from the seuences, plus the four free mutation parameters used in <em>HalperBruno</em>.</p>
</li>
<li><p class="first"><em>avgaafreqsFracTolerated</em> : like <em>avgaafreqsHalpernBruno</em> except that the fixation model is <em>FracTolerated</em>.</p>
</li>
<li><p class="first">Versions of the six models above where the stringency parameter <span class="math">
\(\beta\)</span>
 is fixed to one, and versions where <span class="math">
\(\beta\)</span>
 is treated as a free parameter. These have names like <em>HalpernBruno</em> (when <span class="math">
\(\beta = 1\)</span>
) and like <em>fitbetaHalpernBruno</em> (when <span class="math">
\(\beta\)</span>
 is a free parameter).</p>
</li>
<li><p class="first"><em>GY94_CF3x4_omega-global-one_rates-one</em> : the <a class="reference external" href="http://mbe.oxfordjournals.org/content/11/5/725.full.pdf">Goldman and Yang 1994</a> model with the equilibrium frequencies from the <a class="reference external" href="http://www.plosone.org/article/info%3Adoi/10.1371/journal.pone.0011230">CF3x4</a> method (9 parameters) and a single transition-transversion ratio (<em>kappa</em>) and nonsynonymous-synonymous ration (<em>omega</em>) estimated by maximum likelihood (2 parameters).</p>
</li>
<li><p class="first"><em>GY94_CF3x4_omega-global-one_rates-gamma4</em> : like the <em>GY94_CF3x4_omega-global-one_rates-one</em> model but with the substitution rate drawn from a 4-category discrete gamma distribution (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/7932792">Yang 1994</a>) with the shape parameter estimated by maximum likelihood (adds 1 parameters).</p>
</li>
<li><p class="first"><em>GY94_CF3x4_omega-global-gamma4_rates-one</em> : like the <em>GY94_CF3x4_omega-global-one_rates-one</em> model but with <em>omega</em> drawn from a 4-category discrete gamma distribution (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/7932792">Yang 1994</a>) with the shape parameter estimated by maximum likelihood (adds 1 parameters).</p>
</li>
<li><p class="first"><em>GY94_CF3x4_omega-global-gamma4_rates-gamma4</em> : like the <em>GY94_CF3x4_omega-global-one_rates-one</em> model but with both the rate and <em>omega</em> drawn from their own 4-category discrete gamma distributions (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/7932792">Yang 1994</a>) with the shape parameters estimated by maximum likelihood (adds 2 parameters).</p>
</li>
<li><p class="first"><em>GY94_CF3x4_omega-global-branchlocal_rates-gamma4</em> : like the <em>GY94_CF3x4_omega-global-one_rates-gamma4</em> model but with each branch getting its own <em>omega</em> value optimized by maximum likelihood (adds a number of parameters equal to the number of branch lengths).</p>
</li>
<li><p class="first">Various <em>KOSI07</em> models as described in <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/17400572">Kosiol et al 2007</a> with the equilibrium frequencies estimated using the <em>F</em> method (60 parameters) and the other options as described immediately above for the <em>GY94</em> model.</p>
</li>
</ul>
</blockquote>
<p>A helpful way to view the results is to look at the main summary files created by <tt class="docutils literal">run_analysis.py</tt> in the examples home directory. These files summarize the results for each tree topology.
The results are sorted by the <a class="reference external" href="http://en.wikipedia.org/wiki/Akaike_information_criterion">AIC</a> from best to worst.
Overall, the files contain the following columns:</p>
<blockquote>
<ol class="arabic">
<li><p class="first">Description of the substitution model.</p>
</li>
<li><p class="first">The difference in <a class="reference external" href="http://en.wikipedia.org/wiki/Akaike_information_criterion">AIC</a> from the best model.</p>
</li>
<li><p class="first">The log likelihood</p>
</li>
<li><p class="first">The total number of free parameters in the substitution model. This is the sum of the parameters optimized by maximum likelihood and the parameters estimated empirically from the data.</p>
</li>
<li><p class="first">The total number of parameters in the substitution model optimized by maximum likelihood:</p>
<blockquote>
<ul class="simple">
<li>For experimental models, this is four or five. There are always four optimized mutation rates, and some models also fit the stringency parameter <span class="math">
\(\beta\)</span>
 while others constrain it to one.</li>
<li>For <em>GY94</em> and <em>KOSI07</em> models, this is always at least two, as both <em>omega</em> and <em>kappa</em> are estimated by maximum likelihood. For some of the variants, <em>omega</em> is drawn from discrete gamma categories, which adds an additional parameter (there is now a shape and rate parameter). For some of the variants, the rate is also allowed to be drawn from discrete gamma categories, which also adds another parameter. Finally, for some of the variants <em>omega</em> is different for each branch, which adds a number of free parameters equal to the number of branches.</li>
</ul>
</blockquote>
</li>
<li><p class="first">The total number of parameters in the substitution model that are empirically estimated from the sequence data:</p>
<blockquote>
<ul class="simple">
<li>For experimental models, this is zero.</li>
<li>For <em>GY94</em> models, this is 9 as in the <a class="reference external" href="http://www.plosone.org/article/info%3Adoi/10.1371/journal.pone.0011230">CF3x4</a> model there are a total of three independent frequencies for each of the three codon positions.</li>
<li>For <em>KOSI07</em> models, this is 60 as there are a total of 60 independent frequencies for the 61 non-stop codons.</li>
</ul>
</blockquote>
</li>
</ol>
</blockquote>
<p>Here are the results for optimization of the combined <em>TEM</em> and <em>SHV</em> tree built with the <em>GY94</em> model in <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> (this is the file <tt class="docutils literal">GY94_summary.csv</tt>):</p>
<blockquote>
<pre class="literal-block">
#Summary for tree GY94.
#
#SUBSTITUTION_MODEL, dAIC, LOG_LIKELIHOOD, FREE_PARAMETERS, MAXIMUM_LIKELIHOOD_PARAMETERS, EMPIRICAL_PARAMETERS
fitbetaHalpernBruno, 0, -4020.63, 5, 5, 0
HalpernBruno, 46.44, -4044.85, 4, 4, 0
fitbetaFracTolerated, 77.28, -4059.27, 5, 5, 0
FracTolerated, 85.68, -4064.47, 4, 4, 0
GY94_CF3x4_omega-global-gamma4_rates-gamma4, 392.4, -4208.83, 13, 4, 9
KOSI07_F_omega-global-gamma4_rates-gamma4, 410.703, -4166.98, 64, 4, 60
GY94_CF3x4_omega-global-gamma4_rates-one, 460.9, -4244.08, 12, 3, 9
KOSI07_F_omega-global-gamma4_rates-one, 467.033, -4196.15, 63, 3, 60
GY94_CF3x4_omega-global-one_rates-gamma4, 528.924, -4278.09, 12, 3, 9
KOSI07_F_omega-global-one_rates-gamma4, 551.298, -4238.28, 63, 3, 60
KOSI07_F_omega-global-one_rates-one, 632.878, -4280.07, 62, 2, 60
GY94_CF3x4_omega-global-one_rates-one, 656.184, -4342.72, 11, 2, 9
fitbetaFracToleratedrandomized, 724.54, -4382.9, 5, 5, 0
fitbetaHalpernBrunorandomized, 735.06, -4388.16, 5, 5, 0
avgaafreqsfitbeta_FracTolerated, 820.78, -4371.02, 65, 5, 60
avgaafreqsfitbeta_HalpernBruno, 841.8, -4381.53, 65, 5, 60
avgaafreqs_FracTolerated, 858.04, -4390.65, 64, 4, 60
avgaafreqs_HalpernBruno, 900.68, -4411.97, 64, 4, 60
FracToleratedrandomized, 1264.88, -4654.07, 4, 4, 0
HalpernBrunorandomized, 1474.46, -4758.86, 4, 4, 0
</pre>
</blockquote>
<p>Here are the results for optimization of the combined <em>TEM</em> and <em>SHV</em> tree built with the <em>KOSI07</em> model in <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> (this is the file <tt class="docutils literal">KOSI07_summary.csv</tt>):</p>
<blockquote>
<pre class="literal-block">
#Summary for tree KOSI07.
#
#SUBSTITUTION_MODEL, dAIC, LOG_LIKELIHOOD, FREE_PARAMETERS, MAXIMUM_LIKELIHOOD_PARAMETERS, EMPIRICAL_PARAMETERS
fitbetaHalpernBruno, 0, -4019.94, 5, 5, 0
HalpernBruno, 48.34, -4045.11, 4, 4, 0
fitbetaFracTolerated, 76.06, -4057.97, 5, 5, 0
FracTolerated, 85.6, -4063.74, 4, 4, 0
GY94_CF3x4_omega-global-gamma4_rates-gamma4, 397.965, -4210.92, 13, 4, 9
KOSI07_F_omega-global-gamma4_rates-gamma4, 402.189, -4162.03, 64, 4, 60
KOSI07_F_omega-global-gamma4_rates-one, 455.126, -4189.5, 63, 3, 60
GY94_CF3x4_omega-global-gamma4_rates-one, 464.4, -4245.14, 12, 3, 9
GY94_CF3x4_omega-global-one_rates-gamma4, 527.857, -4276.87, 12, 3, 9
KOSI07_F_omega-global-one_rates-gamma4, 529.827, -4226.85, 63, 3, 60
KOSI07_F_omega-global-one_rates-one, 608.297, -4267.09, 62, 2, 60
GY94_CF3x4_omega-global-one_rates-one, 651.932, -4339.91, 11, 2, 9
fitbetaFracToleratedrandomized, 726.26, -4383.07, 5, 5, 0
fitbetaHalpernBrunorandomized, 736.96, -4388.42, 5, 5, 0
avgaafreqsfitbeta_FracTolerated, 823.74, -4371.81, 65, 5, 60
avgaafreqsfitbeta_HalpernBruno, 844.76, -4382.32, 65, 5, 60
avgaafreqs_FracTolerated, 862.08, -4391.98, 64, 4, 60
avgaafreqs_HalpernBruno, 907.08, -4414.48, 64, 4, 60
FracToleratedrandomized, 1265.14, -4653.51, 4, 4, 0
HalpernBrunorandomized, 1474.08, -4757.98, 4, 4, 0
</pre>
</blockquote>
<p>Here are the results for optimization of the  <em>TEM</em>-only tree built with the <em>GY94</em> model in <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> (this is the file <tt class="docutils literal">GY94_TEM_summary.csv</tt>):</p>
<blockquote>
<pre class="literal-block">
#Summary for tree GY94_TEM.
#
#SUBSTITUTION_MODEL, dAIC, LOG_LIKELIHOOD, FREE_PARAMETERS, MAXIMUM_LIKELIHOOD_PARAMETERS, EMPIRICAL_PARAMETERS
fitbetaHalpernBruno, 0, -2374.29, 5, 5, 0
HalpernBruno, 23.12, -2386.85, 4, 4, 0
fitbetaFracTolerated, 81.8, -2415.19, 5, 5, 0
FracTolerated, 83.6, -2417.09, 4, 4, 0
GY94_CF3x4_omega-global-gamma4_rates-gamma4, 252.239, -2492.41, 13, 4, 9
GY94_CF3x4_omega-global-one_rates-gamma4, 317.554, -2526.07, 12, 3, 9
GY94_CF3x4_omega-global-gamma4_rates-one, 318.468, -2526.52, 12, 3, 9
KOSI07_F_omega-global-gamma4_rates-gamma4, 326.895, -2478.74, 64, 4, 60
KOSI07_F_omega-global-gamma4_rates-one, 394.938, -2513.76, 63, 3, 60
KOSI07_F_omega-global-one_rates-gamma4, 412.027, -2522.3, 63, 3, 60
fitbetaFracToleratedrandomized, 465.8, -2607.19, 5, 5, 0
fitbetaHalpernBrunorandomized, 466.16, -2607.37, 5, 5, 0
GY94_CF3x4_omega-global-one_rates-one, 483.301, -2609.94, 11, 2, 9
KOSI07_F_omega-global-one_rates-one, 556.735, -2595.66, 62, 2, 60
avgaafreqsfitbeta_HalpernBruno, 574.64, -2601.61, 65, 5, 60
avgaafreqsfitbeta_FracTolerated, 577.9, -2603.24, 65, 5, 60
avgaafreqs_HalpernBruno, 609.12, -2619.85, 64, 4, 60
avgaafreqs_FracTolerated, 622.74, -2626.66, 64, 4, 60
FracToleratedrandomized, 976.62, -2863.6, 4, 4, 0
HalpernBrunorandomized, 1007.78, -2879.18, 4, 4, 0
</pre>
</blockquote>
<p>Here are the results for optimization of the  <em>SHV</em>-only tree built with the <em>GY94</em> model in <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> (this is the file <tt class="docutils literal">GY94_SHV_summary.csv</tt>):</p>
<blockquote>
<pre class="literal-block">
#Summary for tree GY94_SHV.
#
#SUBSTITUTION_MODEL, dAIC, LOG_LIKELIHOOD, FREE_PARAMETERS, MAXIMUM_LIKELIHOOD_PARAMETERS, EMPIRICAL_PARAMETERS
fitbetaHalpernBruno, 0, -1728.55, 5, 5, 0
fitbetaFracTolerated, 34.92, -1746.01, 5, 5, 0
HalpernBruno, 106.22, -1782.66, 4, 4, 0
FracTolerated, 116.5, -1787.8, 4, 4, 0
KOSI07_F_omega-global-gamma4_rates-gamma4, 488.954, -1914.03, 64, 4, 60
KOSI07_F_omega-global-one_rates-gamma4, 499.662, -1920.38, 63, 3, 60
GY94_CF3x4_omega-global-gamma4_rates-gamma4, 505.8, -1973.45, 13, 4, 9
GY94_CF3x4_omega-global-one_rates-gamma4, 513.974, -1978.54, 12, 3, 9
KOSI07_F_omega-global-gamma4_rates-one, 555.74, -1948.42, 63, 3, 60
KOSI07_F_omega-global-one_rates-one, 573.58, -1958.34, 62, 2, 60
GY94_CF3x4_omega-global-gamma4_rates-one, 581.526, -2012.31, 12, 3, 9
fitbetaFracToleratedrandomized, 601.68, -2029.39, 5, 5, 0
GY94_CF3x4_omega-global-one_rates-one, 602.573, -2023.84, 11, 2, 9
fitbetaHalpernBrunorandomized, 602.7, -2029.9, 5, 5, 0
avgaafreqsfitbeta_FracTolerated, 711.48, -2024.29, 65, 5, 60
avgaafreqsfitbeta_HalpernBruno, 715.68, -2026.39, 65, 5, 60
avgaafreqs_FracTolerated, 749.82, -2044.46, 64, 4, 60
avgaafreqs_HalpernBruno, 758.76, -2048.93, 64, 4, 60
FracToleratedrandomized, 1047, -2253.05, 4, 4, 0
HalpernBrunorandomized, 1071.2, -2265.15, 4, 4, 0
</pre>
</blockquote>
<p>Here are the results for optimization of the  <em>TEM</em>-only tree built with the <em>KOSI07</em> model in <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> (this is the file <tt class="docutils literal">KOSI07_TEM_summary.csv</tt>):</p>
<blockquote>
<pre class="literal-block">
#Summary for tree KOSI07_TEM.
#
#SUBSTITUTION_MODEL, dAIC, LOG_LIKELIHOOD, FREE_PARAMETERS, MAXIMUM_LIKELIHOOD_PARAMETERS, EMPIRICAL_PARAMETERS
fitbetaHalpernBruno, 0, -2378.53, 5, 5, 0
HalpernBruno, 25.44, -2392.25, 4, 4, 0
fitbetaFracTolerated, 80.64, -2418.85, 5, 5, 0
FracTolerated, 83.68, -2421.37, 4, 4, 0
GY94_CF3x4_omega-global-gamma4_rates-gamma4, 257.61, -2499.34, 13, 4, 9
GY94_CF3x4_omega-global-one_rates-gamma4, 317.651, -2530.36, 12, 3, 9
GY94_CF3x4_omega-global-gamma4_rates-one, 324.101, -2533.58, 12, 3, 9
KOSI07_F_omega-global-gamma4_rates-gamma4, 325.66, -2482.36, 64, 4, 60
KOSI07_F_omega-global-gamma4_rates-one, 393.479, -2517.27, 63, 3, 60
KOSI07_F_omega-global-one_rates-gamma4, 402.405, -2521.73, 63, 3, 60
fitbetaFracToleratedrandomized, 471.96, -2614.51, 5, 5, 0
fitbetaHalpernBrunorandomized, 472.38, -2614.72, 5, 5, 0
GY94_CF3x4_omega-global-one_rates-one, 487.951, -2616.51, 11, 2, 9
KOSI07_F_omega-global-one_rates-one, 550.39, -2596.73, 62, 2, 60
avgaafreqsfitbeta_HalpernBruno, 581.58, -2609.32, 65, 5, 60
avgaafreqsfitbeta_FracTolerated, 584.16, -2610.61, 65, 5, 60
avgaafreqs_HalpernBruno, 617.3, -2628.18, 64, 4, 60
avgaafreqs_FracTolerated, 629.16, -2634.11, 64, 4, 60
FracToleratedrandomized, 980.72, -2869.89, 4, 4, 0
HalpernBrunorandomized, 1014.62, -2886.84, 4, 4, 0
</pre>
</blockquote>
<p>Here are the results for optimization of the  <em>SHV</em>-only tree built with the <em>KOSI07</em> model in <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a> (this is the file <tt class="docutils literal">KOSI07_SHV_summary.csv</tt>):</p>
<blockquote>
<pre class="literal-block">
#Summary for tree KOSI07_SHV.
#
#SUBSTITUTION_MODEL, dAIC, LOG_LIKELIHOOD, FREE_PARAMETERS, MAXIMUM_LIKELIHOOD_PARAMETERS, EMPIRICAL_PARAMETERS
fitbetaHalpernBruno, 0, -1725.36, 5, 5, 0
fitbetaFracTolerated, 34.12, -1742.42, 5, 5, 0
HalpernBruno, 104.38, -1778.55, 4, 4, 0
FracTolerated, 114.6, -1783.66, 4, 4, 0
KOSI07_F_omega-global-gamma4_rates-gamma4, 486.552, -1909.64, 64, 4, 60
KOSI07_F_omega-global-one_rates-gamma4, 491.694, -1913.21, 63, 3, 60
GY94_CF3x4_omega-global-gamma4_rates-gamma4, 497.303, -1966.01, 13, 4, 9
GY94_CF3x4_omega-global-one_rates-gamma4, 501.509, -1969.11, 12, 3, 9
KOSI07_F_omega-global-gamma4_rates-one, 547.601, -1941.16, 63, 3, 60
KOSI07_F_omega-global-one_rates-one, 562.534, -1949.63, 62, 2, 60
GY94_CF3x4_omega-global-gamma4_rates-one, 568, -2002.36, 12, 3, 9
GY94_CF3x4_omega-global-one_rates-one, 586.246, -2012.48, 11, 2, 9
fitbetaFracToleratedrandomized, 597.66, -2024.19, 5, 5, 0
fitbetaHalpernBrunorandomized, 598.68, -2024.7, 5, 5, 0
avgaafreqsfitbeta_FracTolerated, 706.44, -2018.58, 65, 5, 60
avgaafreqsfitbeta_HalpernBruno, 710.94, -2020.83, 65, 5, 60
avgaafreqs_FracTolerated, 745.92, -2039.32, 64, 4, 60
avgaafreqs_HalpernBruno, 755.3, -2044.01, 64, 4, 60
FracToleratedrandomized, 1040.94, -2246.83, 4, 4, 0
HalpernBrunorandomized, 1063.86, -2258.29, 4, 4, 0
</pre>
</blockquote>
<p>Similar files are created in LaTex formatted tables as <tt class="docutils literal">GY94_summary.tex</tt>, <tt class="docutils literal">KOSI07_summary.tex</tt>, etc.</p>
</div>
<div class="section" id="comparison-of-per-site-likelihoods">
<h2><a class="toc-backref" href="#id14">Comparison of per-site likelihoods</a></h2>
<p>The <cite>phyloExpCM</cite> script <tt class="docutils literal">phyloExpCM_SiteLikelihoodComparison.py</tt> was used to compare the per-site likelihoods between the best experimentally informed model (the <em>HalpernBruno</em> model with a fitted <span class="math">
\(\beta\)</span>
) and the best traditional model (the <em>GY94</em> model with gamma-distributed rates and <span class="math">
\(\omega\)</span>
) values for the tree inferred with <em>GY94</em> over the entire lactamase (TEM and SHV) sequences.</p>
<p>This was done by creating the input file <tt class="docutils literal">phyloExpCM_SitelikelihoodComparison_infile.txt</tt> and then running:</p>
<pre class="literal-block">
phyloExpCM_SiteLikelihoodComparison.py phyloExpCM_SiteLikelihoodComparison_infile.txt
</pre>
<p>This created three output files.</p>
<p>The first, <tt class="docutils literal">sitelikelihoods.txt</tt> lists the differences in per-site likelihoods by site.</p>
<p>The next two are the following plots, which use Tukey box plots to show the differences in site likelihoods, categorizing sites by the <a class="reference external" href="http://www.cmbi.ru.nl/hsspsoap/">DSSP webserver</a> computed secondary structures (<em>SS</em>) and relative solvent accessibilities (<em>RSA</em>) in <tt class="docutils literal">1XPB_renumbered.dssp</tt>. Here are these plots:</p>
<div class="figure align-center">
<img alt="sitelikelihoodcomparison_bySS.jpg" src="sitelikelihoodcomparison_bySS.jpg" style="width: 45%;" />
<p class="caption">The comparison of site likelihoods with sites categorized by secondary structure (<tt class="docutils literal">sitelikelihoodcomparison_bySS.pdf</tt>).</p>
</div>
<div class="figure align-center">
<img alt="sitelikelihoodcomparison_byRSA.jpg" src="sitelikelihoodcomparison_byRSA.jpg" style="width: 45%;" />
<p class="caption">The comparison of site likelihoods with sites categorized by relative solvent accessibility (<tt class="docutils literal">sitelikelihoodcomparison_byRSA.pdf</tt>).</p>
</div>
<p>These plots show that the experimental model is modestly better for most sites, but there are a small number of sites for which the experimental model is far worse.</p>
</div>
</div>
<div class="section" id="conclusions">
<h1><a class="toc-backref" href="#id15">Conclusions</a></h1>
<p>The key conclusions that can be drawn from the tables generated by <a class="reference internal" href="#optimizing-the-trees-with-various-substitution-models">Optimizing the trees with various substitution models</a> are as follows:</p>
<ol class="arabic simple">
<li>The evolutionary models that use the experimentally determined amino-acid preferences substantially outperform the other models. The results are slightly better if the fixation probabilities are estimated using the <em>HalpernBruno</em> method.</li>
<li>The superior performance of the models with the experimentally determined preferences is due to the site-specific information. If the preferences are randomized among sites or determined from overall average amino-acid frequencies, the models are far worse than any others.</li>
<li>The <em>GY94</em> and <em>KOSI07</em> models perform similarly -- not as good as the experimentally informed models, but better than the randomized models with the experimental data. Overall <em>KOSI07</em> gives better likelihoods, but not enough so to justify the additional empirically determined equilibrium frequencies according to <a class="reference external" href="http://en.wikipedia.org/wiki/Akaike_information_criterion">AIC</a>. In general, more parameterized versions of the models give better likelihoods, although the improvements are not enough to support the use of branch-local <em>omega</em> values.</li>
<li>These results are robust to the exact sequences and model used to the build the tree. They are true for trees built using <em>GY94</em> or <em>KOSI07</em>, and for trees with the combined <em>TEM</em> and <em>SHV</em> sequences or for trees with just one sequence subset alone.</li>
<li>Incorporation of the stringency parameter <span class="math">
\(\beta\)</span>
 leads to improvements in the experimentally informed models. The fit values of <span class="math">
\(\beta\)</span>
 are greater than one, suggesting that real evolution is more stringent than the deep mutational scanning.</li>
</ol>
<p><strong>Overall, these results demonstrate that using the experimentally determined amino-acid preferences improves phylogenetic fit over existing approaches.</strong></p>
</div>
</div>
</body>
</html>
